// @plur-ai/core/dist/index.js
import {
  atomicWrite,
  embeddingSearch,
  engramSearchText,
  ftsScore,
  ftsTokenize,
  getSyncStatus,
  searchEngrams,
  sync,
  withLock
} from "./chunk-KMVQYBNP.js";
import "./chunk-2ZDO52B4.js";

// src/index.ts
import * as fs3 from "fs";
import yaml5 from "js-yaml";

// src/storage.ts
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import { homedir } from "os";
function detectPlurStorage(explicitPath) {
  const root = explicitPath || process.env.PLUR_PATH || join(homedir(), ".plur");
  if (!existsSync(root)) mkdirSync(root, { recursive: true });
  const packsDir = join(root, "packs");
  if (!existsSync(packsDir)) mkdirSync(packsDir, { recursive: true });
  return {
    root,
    engrams: join(root, "engrams.yaml"),
    episodes: join(root, "episodes.yaml"),
    candidates: join(root, "candidates.yaml"),
    packs: packsDir,
    exchange: join(root, "exchange"),
    config: join(root, "config.yaml"),
    db: join(root, "engrams.db")
  };
}

// src/storage-indexed.ts
import { existsSync as existsSync3 } from "fs";
import { createRequire } from "module";

// src/engrams.ts
import * as fs from "fs";
import * as yaml from "js-yaml";

// src/schemas/engram.ts
import { z } from "zod";
var ActivationSchema = z.object({
  retrieval_strength: z.number().min(0).max(1),
  storage_strength: z.number().min(0).max(1),
  frequency: z.number().int().min(0),
  last_accessed: z.string()
});
var KnowledgeTypeSchema = z.object({
  memory_class: z.enum(["semantic", "episodic", "procedural", "metacognitive"]),
  cognitive_level: z.enum(["remember", "understand", "apply", "analyze", "evaluate", "create"])
});
var KnowledgeAnchorSchema = z.object({
  path: z.string(),
  relevance: z.enum(["primary", "supporting", "example"]).default("supporting"),
  snippet: z.string().max(200).optional(),
  snippet_extracted_at: z.string().optional()
});
var AssociationSchema = z.object({
  target_type: z.enum(["engram", "document"]),
  target: z.string(),
  strength: z.number().min(0).max(0.95),
  type: z.enum(["semantic", "temporal", "causal", "co_accessed"]),
  updated_at: z.string().optional()
});
var DualCodingSchema = z.object({
  example: z.string().optional(),
  analogy: z.string().optional()
}).refine(
  (d) => d.example || d.analogy,
  "At least one of example or analogy must be provided"
);
var RelationsSchema = z.object({
  broader: z.array(z.string()).default([]),
  narrower: z.array(z.string()).default([]),
  related: z.array(z.string()).default([]),
  conflicts: z.array(z.string()).default([])
});
var ProvenanceSchema = z.object({
  origin: z.string(),
  chain: z.array(z.string()).default([]),
  signature: z.string().nullable().default(null),
  license: z.string().default("cc-by-sa-4.0")
});
var FeedbackSignalsSchema = z.object({
  positive: z.number().int().default(0),
  negative: z.number().int().default(0),
  neutral: z.number().int().default(0)
});
var EntityRefSchema = z.object({
  name: z.string(),
  type: z.enum([
    "person",
    "organization",
    "technology",
    "concept",
    "project",
    "tool",
    "place",
    "event",
    "standard",
    "other"
  ]),
  uri: z.string().url().optional()
});
var TemporalSchema = z.object({
  learned_at: z.string(),
  valid_from: z.string().optional(),
  valid_until: z.string().optional(),
  ingested_at: z.string().optional()
});
var UsageStatsSchema = z.object({
  injections: z.number().int().default(0),
  hits: z.number().int().default(0),
  misses: z.number().int().default(0),
  last_hit_at: z.string().optional()
});
var EpisodicFieldsSchema = z.object({
  emotional_weight: z.number().int().min(1).max(10).default(5),
  confidence: z.number().int().min(1).max(10).default(5),
  trigger_context: z.string().optional(),
  journal_ref: z.string().optional()
});
var ExchangeMetadataSchema = z.object({
  fitness_score: z.number().min(0).max(1).optional(),
  environmental_diversity: z.number().int().default(0),
  adoption_count: z.number().int().default(0),
  contradiction_rate: z.number().min(0).max(1).default(0)
});
var EngramSchema = z.object({
  // Identity
  id: z.string().regex(/^(ENG|ABS|META)-[A-Za-z0-9-]+$/),
  version: z.number().int().min(1).default(2),
  status: z.enum(["active", "dormant", "retired", "candidate"]),
  consolidated: z.boolean().default(false),
  type: z.enum(["behavioral", "terminological", "procedural", "architectural"]),
  scope: z.string(),
  visibility: z.enum(["private", "public", "template"]).default("private"),
  // Content
  statement: z.string().min(1),
  rationale: z.string().optional(),
  contraindications: z.array(z.string()).optional(),
  // Lineage
  source: z.string().optional(),
  source_patterns: z.array(z.string()).optional(),
  derivation_count: z.number().int().min(0).default(1),
  pack: z.string().nullable().default(null),
  abstract: z.string().nullable().default(null),
  derived_from: z.string().nullable().default(null),
  // Classification
  knowledge_type: KnowledgeTypeSchema.optional(),
  domain: z.string().optional(),
  tags: z.array(z.string()).default([]),
  // Activation (ACT-R model)
  activation: ActivationSchema.default({
    retrieval_strength: 0.7,
    storage_strength: 1,
    frequency: 0,
    last_accessed: (/* @__PURE__ */ new Date()).toISOString().slice(0, 10)
  }),
  // Relations & grounding
  relations: RelationsSchema.optional(),
  associations: z.array(AssociationSchema).default([]),
  knowledge_anchors: z.array(KnowledgeAnchorSchema).default([]),
  dual_coding: DualCodingSchema.optional(),
  // Provenance
  provenance: ProvenanceSchema.optional(),
  // Feedback
  feedback_signals: FeedbackSignalsSchema.default({ positive: 0, negative: 0, neutral: 0 }),
  // === NEW OPTIONAL FIELDS (v2.1) ===
  /** Typed entity references extracted from statement. Enables graph queries. */
  entities: z.array(EntityRefSchema).optional(),
  /** Temporal validity window. When is this knowledge true? */
  temporal: TemporalSchema.optional(),
  /** Automatic usage tracking. Injections, hits, misses. */
  usage: UsageStatsSchema.optional(),
  /** Episodic context: emotional weight, confidence, trigger. */
  episodic: EpisodicFieldsSchema.optional(),
  /** Exchange marketplace metadata: fitness, adoption, diversity. */
  exchange: ExchangeMetadataSchema.optional(),
  /** Extensible key-value data for domain-specific fields. */
  structured_data: z.record(z.string(), z.unknown()).optional(),
  /** Polarity classification: 'do' for directives, 'dont' for prohibitions, null for unclassified. */
  polarity: z.enum(["do", "dont"]).nullable().default(null)
});

// src/schemas/pack.ts
import { z as z2 } from "zod";
var PackManifestSchema = z2.object({
  name: z2.string(),
  version: z2.string(),
  description: z2.string().optional(),
  creator: z2.string().optional(),
  license: z2.string().default("cc-by-sa-4.0"),
  tags: z2.array(z2.string()).default([]),
  metadata: z2.object({
    id: z2.string().optional(),
    injection_policy: z2.enum(["on_match", "on_request", "always"]).default("on_match"),
    match_terms: z2.array(z2.string()).default([]),
    domain: z2.string().optional(),
    engram_count: z2.number().optional()
  }).optional(),
  "x-datacore": z2.object({
    id: z2.string(),
    injection_policy: z2.enum(["on_match", "on_request"]),
    match_terms: z2.array(z2.string()).default([]),
    domain: z2.string().optional(),
    engram_count: z2.number().int().min(0)
  }).optional()
});

// src/logger.ts
var level = process.env.PLUR_LOG_LEVEL || "warning";
var levels = { debug: 0, info: 1, warning: 2, error: 3 };
var threshold = levels[level] ?? 2;
var logger = {
  debug: (...args) => {
    if (threshold <= 0) console.error("[plur:debug]", ...args);
  },
  info: (...args) => {
    if (threshold <= 1) console.error("[plur:info]", ...args);
  },
  warning: (...args) => {
    if (threshold <= 2) console.error("[plur:warning]", ...args);
  },
  error: (...args) => {
    if (threshold <= 3) console.error("[plur:error]", ...args);
  }
};

// src/engrams.ts
function loadEngrams(filePath) {
  if (!fs.existsSync(filePath)) return [];
  try {
    const raw = yaml.load(fs.readFileSync(filePath, "utf8"));
    if (!raw?.engrams || !Array.isArray(raw.engrams)) return [];
    const valid = [];
    let skipped = 0;
    for (const entry of raw.engrams) {
      const result = EngramSchema.safeParse(entry);
      if (result.success) valid.push(result.data);
      else skipped++;
    }
    if (skipped > 0) logger.warning(`Skipped ${skipped} invalid engram(s) in ${filePath}`);
    return valid;
  } catch (err) {
    logger.error(`Failed to parse engrams file ${filePath}: ${err}`);
    return [];
  }
}
function saveEngrams(filePath, engrams) {
  const content = yaml.dump({ engrams }, { lineWidth: 120, noRefs: true, quotingType: '"' });
  atomicWrite(filePath, content);
}
function parseSkillMdFrontmatter(filePath) {
  const content = fs.readFileSync(filePath, "utf8");
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) throw new Error(`No frontmatter found in ${filePath}`);
  return yaml.load(match[1]);
}
function loadPack(packDir) {
  const skillMdPath = `${packDir}/SKILL.md`;
  const manifestYamlPath = `${packDir}/manifest.yaml`;
  const engramsPath = `${packDir}/engrams.yaml`;
  let rawManifest;
  if (fs.existsSync(skillMdPath)) {
    rawManifest = parseSkillMdFrontmatter(skillMdPath);
  } else if (fs.existsSync(manifestYamlPath)) {
    rawManifest = yaml.load(fs.readFileSync(manifestYamlPath, "utf8"));
  } else {
    throw new Error(`No SKILL.md or manifest.yaml found in ${packDir}`);
  }
  const manifest = PackManifestSchema.parse(rawManifest);
  const engrams = loadEngrams(engramsPath);
  return { manifest, engrams };
}
function loadAllPacks(packsDir) {
  if (!fs.existsSync(packsDir)) return [];
  const packs = [];
  for (const entry of fs.readdirSync(packsDir)) {
    const packDir = `${packsDir}/${entry}`;
    if (!fs.statSync(packDir).isDirectory()) continue;
    if (!fs.existsSync(`${packDir}/SKILL.md`) && !fs.existsSync(`${packDir}/manifest.yaml`)) continue;
    try {
      packs.push(loadPack(packDir));
    } catch (err) {
      logger.warning(`Failed to load pack ${entry}: ${err}`);
    }
  }
  return packs;
}
function storePrefix(scope) {
  const parts = scope.split(/[:\-_./]/).filter(Boolean);
  if (parts.length >= 2) {
    const p2 = parts[1];
    return (parts[0][0] + p2[0] + (p2[1] || p2[0])).toUpperCase();
  }
  const w = parts[0] || scope;
  if (w.length >= 3) return (w[0] + w[Math.floor(w.length / 2)] + w[w.length - 1]).toUpperCase();
  return (w[0] + (w[1] || w[0]) + (w[2] || w[0])).toUpperCase();
}
function generateEngramId(existing) {
  const now = /* @__PURE__ */ new Date();
  const date = now.toISOString().slice(0, 10).replace(/-/g, "");
  const prefix = `ENG-${date.slice(0, 4)}-${date.slice(4)}-`;
  const existingNums = existing.filter((e) => e.id.startsWith(prefix)).map((e) => parseInt(e.id.slice(prefix.length), 10)).filter((n) => !isNaN(n));
  const next = existingNums.length > 0 ? Math.max(...existingNums) + 1 : 1;
  return `${prefix}${String(next).padStart(3, "0")}`;
}

// src/storage-indexed.ts
var require2 = createRequire(import.meta.url);
var Database = null;
function getDatabase() {
  if (!Database) {
    try {
      Database = require2("better-sqlite3");
    } catch {
      throw new Error(
        "better-sqlite3 is required for index: true. Install it with: npm install better-sqlite3"
      );
    }
  }
  return Database;
}
var IndexedStorage = class {
  dbPath;
  engramsPath;
  stores;
  db = null;
  constructor(engramsPath, dbPath, stores) {
    this.engramsPath = engramsPath;
    this.dbPath = dbPath;
    this.stores = stores ?? [];
  }
  getDb() {
    if (!this.db) {
      const DB = getDatabase();
      this.db = new DB(this.dbPath);
      this.db.pragma("journal_mode = WAL");
      this.db.exec(`
        CREATE TABLE IF NOT EXISTS engrams (
          id TEXT PRIMARY KEY,
          status TEXT NOT NULL,
          scope TEXT NOT NULL,
          domain TEXT,
          last_accessed TEXT,
          data TEXT NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_status ON engrams(status);
        CREATE INDEX IF NOT EXISTS idx_scope ON engrams(scope);
        CREATE INDEX IF NOT EXISTS idx_domain ON engrams(domain);
      `);
      try {
        this.db.exec("ALTER TABLE engrams ADD COLUMN source TEXT NOT NULL DEFAULT 'primary'");
      } catch {
      }
      this.db.exec("CREATE INDEX IF NOT EXISTS idx_source ON engrams(source)");
    }
    return this.db;
  }
  /** Load all engrams from SQLite index. Auto-rebuilds if db missing. */
  loadAll() {
    if (!existsSync3(this.dbPath)) {
      this.reindex();
    }
    const db = this.getDb();
    const rows = db.prepare("SELECT data FROM engrams").all();
    return rows.map((r) => JSON.parse(r.data));
  }
  /** Load engrams with SQL-level filtering. */
  loadFiltered(filter) {
    if (!existsSync3(this.dbPath)) {
      this.reindex();
    }
    const db = this.getDb();
    const conditions = [];
    const params = [];
    if (filter.status) {
      conditions.push("status = ?");
      params.push(filter.status);
    }
    if (filter.scope) {
      conditions.push("(scope = 'global' OR scope = ? OR scope LIKE ? || '%')");
      params.push(filter.scope, filter.scope);
    }
    if (filter.domain) {
      conditions.push("domain LIKE ? || '%'");
      params.push(filter.domain);
    }
    const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    const rows = db.prepare(`SELECT data FROM engrams ${where}`).all(...params);
    return rows.map((r) => JSON.parse(r.data));
  }
  /** Count engrams with optional status filter. */
  count(filter) {
    if (!existsSync3(this.dbPath)) {
      this.reindex();
    }
    const db = this.getDb();
    if (filter?.status) {
      return db.prepare("SELECT COUNT(*) as c FROM engrams WHERE status = ?").get(filter.status).c;
    }
    return db.prepare("SELECT COUNT(*) as c FROM engrams").get().c;
  }
  /** Sync SQLite index from YAML source of truth (primary + all stores). */
  syncFromYaml() {
    const db = this.getDb();
    const upsert = db.prepare(`
      INSERT OR REPLACE INTO engrams (id, status, scope, domain, last_accessed, data, source)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `);
    const allSyncedIds = /* @__PURE__ */ new Set();
    const validSources = /* @__PURE__ */ new Set(["primary"]);
    const tx = db.transaction(() => {
      const primaryEngrams = loadEngrams(this.engramsPath);
      for (const e of primaryEngrams) {
        upsert.run(e.id, e.status, e.scope, e.domain ?? null, e.activation.last_accessed, JSON.stringify(e), "primary");
        allSyncedIds.add(e.id);
      }
      for (const store of this.stores) {
        validSources.add(store.path);
        const storeEngrams = loadEngrams(store.path);
        const prefix = storePrefix(store.scope);
        for (const e of storeEngrams) {
          if (e.scope !== "global" && e.scope !== store.scope && !e.scope.startsWith(store.scope)) {
            continue;
          }
          const nsId = e.id.replace(/^(ENG|ABS|META)-/, `$1-${prefix}-`);
          const scope = e.scope === "global" ? store.scope : e.scope;
          upsert.run(nsId, e.status, scope, e.domain ?? null, e.activation.last_accessed, JSON.stringify({ ...e, id: nsId, scope }), store.path);
          allSyncedIds.add(nsId);
        }
      }
      const dbRows = db.prepare("SELECT id, source FROM engrams").all();
      const deleteStmt = db.prepare("DELETE FROM engrams WHERE id = ?");
      for (const row of dbRows) {
        if (!allSyncedIds.has(row.id)) {
          deleteStmt.run(row.id);
        }
      }
    });
    tx();
  }
  /** Drop and rebuild the entire index from YAML. */
  reindex() {
    this.close();
    const db = this.getDb();
    db.exec("DELETE FROM engrams");
    this.syncFromYaml();
  }
  /** Close the database connection. */
  close() {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }
};

// src/config.ts
import { existsSync as existsSync4, readFileSync as readFileSync2 } from "fs";
import yaml2 from "js-yaml";

// src/schemas/config.ts
import { z as z3 } from "zod";
var StoreEntrySchema = z3.object({
  path: z3.string(),
  scope: z3.string(),
  shared: z3.boolean().default(false),
  readonly: z3.boolean().default(false)
});
var PlurConfigSchema = z3.object({
  auto_learn: z3.boolean().default(true),
  auto_capture: z3.boolean().default(true),
  injection_budget: z3.number().default(2e3),
  decay_enabled: z3.boolean().default(true),
  decay_threshold: z3.number().default(0.15),
  packs: z3.array(z3.string()).default([]),
  injection: z3.object({
    spread_cap: z3.number().default(3),
    spread_budget: z3.number().default(480),
    co_access: z3.boolean().default(true)
  }).default({}),
  allow_secrets: z3.boolean().default(false),
  index: z3.boolean().default(true),
  stores: z3.array(StoreEntrySchema).default([])
}).partial();

// src/config.ts
function loadConfig(configPath) {
  if (!existsSync4(configPath)) return PlurConfigSchema.parse({});
  try {
    const raw = yaml2.load(readFileSync2(configPath, "utf8"));
    return PlurConfigSchema.parse(raw ?? {});
  } catch {
    return PlurConfigSchema.parse({});
  }
}

// src/decay.ts
var DECAY_RATE = 0.05;
var FLOOR = 0.05;
var MS_PER_DAY = 864e5;
function decayedStrength(retrievalStrength, daysSinceAccess, lambda = DECAY_RATE) {
  return FLOOR + (retrievalStrength - FLOOR) * Math.exp(-lambda * daysSinceAccess);
}
function daysSince(lastAccessed, now) {
  const last = new Date(lastAccessed);
  const current = now || /* @__PURE__ */ new Date();
  return Math.max(0, Math.floor((current.getTime() - last.getTime()) / MS_PER_DAY));
}
function reactivate(currentStrength) {
  return Math.min(1, currentStrength + 0.1);
}
function decayedCoAccessStrength(strength, daysSinceUpdate, lambda = 0.01) {
  const floor = 0.02;
  return floor + (strength - floor) * Math.exp(-lambda * daysSinceUpdate);
}

// src/polarity.ts
var DONT_PATTERNS = [
  /\bnever\b/i,
  /\bdo\s+not\b/i,
  /\bdon'?t\b/i,
  /\bavoid\b/i,
  /\bmust\s+not\b/i,
  /\bshould\s+not\b/i,
  /\bNOT\b.*?\b(?:include|use|run|start|create|add|send|describe|reference|assume)\b/,
  /\bstop\b.*?\b(?:generating|iterating|doing|running|creating)\b/i
];
function classifyPolarity(statement) {
  if (!statement || statement.length < 5) return null;
  const dotIndex = statement.indexOf(".");
  const firstSentence = statement.slice(0, dotIndex > 0 ? Math.min(dotIndex, 200) : 200);
  for (const pattern of DONT_PATTERNS) {
    if (pattern.test(firstSentence)) return "dont";
  }
  return null;
}

// src/confidence.ts
function computeConfidence(input) {
  const fb = input.feedback_signals ?? { positive: 0, negative: 0, neutral: 0 };
  const total = fb.positive + fb.negative + fb.neutral;
  if (total === 0) return 0.5;
  const netRatio = (fb.positive - fb.negative) / total;
  const dampening = 1 - 1 / (total + 1);
  const adjustedRatio = netRatio * dampening;
  const steepness = 2;
  const base = 1 / (1 + Math.exp(-steepness * adjustedRatio));
  const consolidationBonus = input.consolidated ? 0.05 : 0;
  return Math.min(1, Math.max(0, base + consolidationBonus));
}
function computeMetaConfidence(evidenceCount, domainCount, structuralDepth, validationRatio) {
  const evidenceSignal = Math.min(evidenceCount / 5, 1) * 0.25;
  const domainSignal = Math.min(domainCount / 3, 1) * 0.35;
  const depthSignal = Math.min(structuralDepth / 3, 1) * 0.2;
  const validationSignal = validationRatio * 0.2;
  return evidenceSignal + domainSignal + depthSignal + validationSignal;
}
function confidenceBand(score) {
  if (score >= 0.7) return "high";
  if (score >= 0.4) return "medium";
  return "low";
}

// src/inject.ts
var DEFAULT_MAX_TOKENS = 8e3;
var DEFAULT_MIN_RELEVANCE = 0.3;
var MAX_PER_PACK = 5;
var MAX_PER_DOMAIN = 10;
var DIP19_CONSIDER_MAX = 5;
var DIP19_CONSIDER_BUDGET = 200;
function getPackMetadata(manifest) {
  const meta = manifest["x-datacore"] || manifest.metadata;
  return {
    injection_policy: meta?.injection_policy ?? "on_match",
    match_terms: meta?.match_terms ?? []
  };
}
function estimateTokens(engram) {
  const { keyword_match: _km, raw_score: _rs, score: _s, associations: _a, ...wire } = engram;
  const serialized = JSON.stringify(wire);
  return Math.ceil(serialized.length / 4);
}
function tokenize(text) {
  return new Set(text.toLowerCase().split(/\W+/).filter((w) => w.length > 2));
}
function anchorBoost(engram, taskWords) {
  if (!engram.knowledge_anchors?.length) return 0;
  const threshold2 = taskWords.size <= 1 ? 1 : 2;
  let boost = 0;
  for (const anchor of engram.knowledge_anchors) {
    if (!anchor.snippet) continue;
    const snippetWords = tokenize(anchor.snippet);
    let overlap = 0;
    for (const word of taskWords) {
      if (snippetWords.has(word)) overlap++;
    }
    if (overlap >= threshold2) boost += 0.5;
  }
  return Math.min(boost, 2);
}
function flattenRelations(engram) {
  if (!engram.relations) return [];
  const associations = [];
  for (const id of engram.relations.broader) {
    associations.push({ target_type: "engram", target: id, type: "semantic", strength: 0.5 });
  }
  for (const id of engram.relations.narrower) {
    associations.push({ target_type: "engram", target: id, type: "semantic", strength: 0.5 });
  }
  for (const id of engram.relations.related) {
    associations.push({ target_type: "engram", target: id, type: "semantic", strength: 0.5 });
  }
  return associations;
}
function stripAssociations(engram) {
  const { associations: _, ...rest } = engram;
  return rest;
}
function stripScoring(engram) {
  const { keyword_match: _, raw_score: _r, score: _s, ...rest } = engram;
  return { ...rest, confidence_score: computeConfidence(engram) };
}
function scoreEngram(engram, promptLower, promptWords, packMatchTerms, scopeFilter, isPack) {
  if (scopeFilter) {
    if (scopeFilter === "global") {
      if (engram.scope !== "global") return 0;
    } else if (!engram.scope.startsWith(scopeFilter) && engram.scope !== "global") {
      return 0;
    }
  }
  let termHits = 0;
  for (const term of packMatchTerms) {
    if (promptLower.includes(term.toLowerCase())) termHits++;
  }
  for (const tag of engram.tags) {
    if (promptWords.has(tag.toLowerCase())) termHits++;
  }
  if (engram.domain) {
    for (const part of engram.domain.split(/[./]/)) {
      if (promptWords.has(part.toLowerCase())) termHits++;
    }
  }
  const statementWords = new Set(engram.statement.toLowerCase().split(/\W+/).filter((w) => w.length > 2));
  for (const word of promptWords) {
    if (statementWords.has(word)) termHits += 0.5;
  }
  if (termHits === 0) return 0;
  const rs = isPack ? engram.activation.retrieval_strength : decayedStrength(engram.activation.retrieval_strength, daysSince(engram.activation.last_accessed));
  let score = termHits * rs;
  const feedback = engram.feedback_signals;
  if (feedback) {
    const netFeedback = feedback.positive - feedback.negative;
    if (netFeedback > 0) score *= 1 + Math.min(netFeedback * 0.05, 0.3);
    else if (netFeedback < 0) score *= Math.max(1 + netFeedback * 0.1, 0.5);
  }
  if (engram.consolidated) score *= 1.1;
  const emotionalWeight = engram.episodic?.emotional_weight ?? 5;
  score *= 1 + (emotionalWeight - 5) * 0.04;
  return score;
}
function fillTokenBudget(scored, maxTokens) {
  const result = [];
  const packCounts = /* @__PURE__ */ new Map();
  const domainCounts = /* @__PURE__ */ new Map();
  let tokensUsed = 0;
  for (const engram of scored) {
    const cost = estimateTokens(engram);
    if (tokensUsed + cost > maxTokens) continue;
    const pack = engram.pack ?? "__personal__";
    const packCount = packCounts.get(pack) ?? 0;
    if (packCount >= MAX_PER_PACK && pack !== "__personal__") continue;
    const domain = engram.domain ?? "__none__";
    const topDomain = domain.split(".")[0];
    const domainCount = domainCounts.get(topDomain) ?? 0;
    if (domainCount >= MAX_PER_DOMAIN) continue;
    result.push(engram);
    tokensUsed += cost;
    packCounts.set(pack, packCount + 1);
    domainCounts.set(topDomain, domainCount + 1);
  }
  return { selected: result, tokens_used: tokensUsed };
}
function selectAndSpread(ctx, personalEngrams, packs, config, embeddingBoosts) {
  const spreadCap = config?.spread_cap ?? 3;
  const spreadBudget = config?.spread_budget ?? 480;
  const promptLower = ctx.prompt.toLowerCase();
  const promptWords = new Set(promptLower.split(/\W+/).filter((w) => w.length > 2));
  const maxTokens = ctx.maxTokens ?? DEFAULT_MAX_TOKENS;
  const minRelevance = ctx.minRelevance ?? DEFAULT_MIN_RELEVANCE;
  const today = (/* @__PURE__ */ new Date()).toISOString().slice(0, 10);
  const engramMap = /* @__PURE__ */ new Map();
  const scored = [];
  for (const engram of personalEngrams) {
    if (engram.status !== "active") continue;
    if (engram.temporal?.valid_until && engram.temporal.valid_until < today) continue;
    if (engram.temporal?.valid_from && engram.temporal.valid_from > today) continue;
    engramMap.set(engram.id, engram);
    let raw = scoreEngram(engram, promptLower, promptWords, [], ctx.scope, false);
    const embBoost = embeddingBoosts?.get(engram.id) ?? 0;
    if (raw === 0 && embBoost > 0.3) {
      raw = embBoost * 2;
    } else if (raw > 0 && embBoost > 0) {
      raw += embBoost;
    }
    if (raw > 0) {
      scored.push({ ...engram, keyword_match: raw, raw_score: raw, score: raw });
    }
  }
  for (const pack of packs) {
    const packMeta = getPackMetadata(pack.manifest);
    if (packMeta.injection_policy === "on_request") continue;
    const matchTerms = packMeta.match_terms;
    for (const engram of pack.engrams) {
      if (engram.status !== "active") continue;
      if (engram.temporal?.valid_until && engram.temporal.valid_until < today) continue;
      if (engram.temporal?.valid_from && engram.temporal.valid_from > today) continue;
      engramMap.set(engram.id, engram);
      let raw = scoreEngram(engram, promptLower, promptWords, matchTerms, ctx.scope, true);
      const embBoost = embeddingBoosts?.get(engram.id) ?? 0;
      if (raw === 0 && embBoost > 0.3) {
        raw = embBoost * 2;
      } else if (raw > 0 && embBoost > 0) {
        raw += embBoost;
      }
      if (raw > 0) {
        scored.push({ ...engram, keyword_match: raw, raw_score: raw, score: raw });
      }
    }
  }
  const maxKm = Math.max(...scored.map((e) => e.keyword_match), 1);
  for (const e of scored) {
    e.keyword_match = e.keyword_match / maxKm * 10;
  }
  const aBoosts = /* @__PURE__ */ new Map();
  for (const e of scored) {
    const aBoost = anchorBoost(e, promptWords);
    aBoosts.set(e.id, aBoost);
    e.score = e.keyword_match + aBoost;
  }
  const filtered = scored.filter((s) => s.score >= minRelevance);
  filtered.sort((a, b) => b.score - a.score);
  const { selected: directives, tokens_used: directiveTokens } = fillTokenBudget(filtered, maxTokens);
  const directiveIds = new Set(directives.map((e) => e.id));
  const directivePackCounts = /* @__PURE__ */ new Map();
  for (const e of directives) {
    const pack = e.pack ?? "__personal__";
    directivePackCounts.set(pack, (directivePackCounts.get(pack) ?? 0) + 1);
  }
  const dip19Remainder = filtered.filter((e) => {
    if (directiveIds.has(e.id)) return false;
    const pack = e.pack ?? "__personal__";
    if (pack !== "__personal__" && (directivePackCounts.get(pack) ?? 0) >= MAX_PER_PACK) return false;
    return true;
  });
  const { selected: dip19Consider } = fillTokenBudget(
    dip19Remainder,
    DIP19_CONSIDER_BUDGET
  );
  const dip19Pool = dip19Consider.slice(0, DIP19_CONSIDER_MAX);
  const dip19PoolTokens = dip19Pool.reduce((acc, e) => acc + estimateTokens(e), 0);
  if (directives.length === 0 && dip19Pool.length === 0) {
    return {
      directives: [],
      constraints: [],
      consider: [],
      tokens_used: { directives: 0, consider: 0 }
    };
  }
  const maxFirstPass = Math.max(...directives.map((e) => e.score), 1);
  const visited = new Set(directives.map((e) => e.id));
  for (const e of dip19Pool) visited.add(e.id);
  const spreadCandidates = [];
  let spreadTokens = 0;
  for (const directive of directives) {
    const assocs = directive.associations?.length ? directive.associations : flattenRelations(directive);
    for (const assoc of assocs) {
      if (assoc.target_type !== "engram") continue;
      if (visited.has(assoc.target)) continue;
      const target = engramMap.get(assoc.target);
      if (!target || target.status !== "active") continue;
      const effectiveStrength = assoc.type === "co_accessed" && assoc.updated_at ? decayedCoAccessStrength(assoc.strength, daysSince(assoc.updated_at)) : assoc.strength;
      if (effectiveStrength <= 0) continue;
      const spreadScore = directive.score / maxFirstPass * effectiveStrength;
      if (spreadScore < minRelevance * 0.5) continue;
      const spreadEngram = {
        ...target,
        keyword_match: 0,
        raw_score: 0,
        score: spreadScore
      };
      const cost = estimateTokens(spreadEngram);
      if (spreadTokens + cost > spreadBudget) continue;
      if (spreadCandidates.length >= spreadCap) break;
      spreadCandidates.push(spreadEngram);
      spreadTokens += cost;
      visited.add(assoc.target);
    }
  }
  const allConsider = [...dip19Pool, ...spreadCandidates];
  const agentDirectives = directives.map(stripAssociations);
  const agentConsider = allConsider.map(stripAssociations);
  const wireAll = agentDirectives.map(stripScoring);
  const wireConsider = agentConsider.map(stripScoring);
  const wireDirectives = [];
  const wireConstraints = [];
  for (const wire of wireAll) {
    const polarity = wire.polarity ?? classifyPolarity(wire.statement);
    if (polarity === "dont") {
      wireConstraints.push(wire);
    } else {
      wireDirectives.push(wire);
    }
  }
  const considerTokens = dip19PoolTokens + spreadTokens;
  return {
    directives: wireDirectives,
    constraints: wireConstraints,
    consider: wireConsider,
    tokens_used: { directives: directiveTokens, consider: considerTokens }
  };
}

// src/episodes.ts
import { existsSync as existsSync5, readFileSync as readFileSync3 } from "fs";
import yaml3 from "js-yaml";
function generateEpisodeId() {
  const ts = Date.now();
  const rand = Math.random().toString(36).slice(2, 6);
  return `EP-${ts}-${rand}`;
}
function captureEpisode(path2, summary, context) {
  const episodes = loadEpisodes(path2);
  const episode = {
    id: generateEpisodeId(),
    summary,
    agent: context?.agent,
    channel: context?.channel,
    session_id: context?.session_id,
    tags: context?.tags,
    timestamp: (/* @__PURE__ */ new Date()).toISOString()
  };
  episodes.push(episode);
  atomicWrite(path2, yaml3.dump(episodes, { lineWidth: 120, noRefs: true }));
  return episode;
}
function queryTimeline(path2, query) {
  let episodes = loadEpisodes(path2);
  if (query?.since) episodes = episodes.filter((e) => new Date(e.timestamp) >= query.since);
  if (query?.until) episodes = episodes.filter((e) => new Date(e.timestamp) <= query.until);
  if (query?.agent) episodes = episodes.filter((e) => e.agent === query.agent);
  if (query?.channel) episodes = episodes.filter((e) => e.channel === query.channel);
  if (query?.search) {
    const terms = query.search.toLowerCase().split(/\s+/);
    episodes = episodes.filter((e) => terms.some((t) => e.summary.toLowerCase().includes(t)));
  }
  return episodes;
}
function loadEpisodes(path2) {
  if (!existsSync5(path2)) return [];
  try {
    const raw = yaml3.load(readFileSync3(path2, "utf8"));
    return Array.isArray(raw) ? raw : [];
  } catch {
    return [];
  }
}

// src/conflict.ts
function detectConflicts(newEngram, existing, threshold2 = 0.4) {
  const newScope = newEngram.scope || "global";
  const newTokens = ftsTokenize(newEngram.statement);
  if (newTokens.length === 0) return [];
  return existing.filter((e) => {
    if (e.status !== "active") return false;
    if ((e.scope || "global") !== newScope) return false;
    const score = ftsScore(e, newTokens);
    return score >= threshold2;
  });
}

// src/agentic-search.ts
async function agenticSearch(engrams, query, limit, llm) {
  const candidates = searchEngrams(engrams, query, Math.min(30, engrams.length));
  if (candidates.length === 0) return [];
  if (candidates.length <= limit) {
    return agenticRerank(candidates, query, limit, llm);
  }
  return agenticRerank(candidates, query, limit, llm);
}
async function agenticRerank(candidates, query, limit, llm) {
  const numbered = candidates.map((e, i) => {
    const extra = [];
    if (e.entities?.length) extra.push(`entities: ${e.entities.map((x) => x.name).join(", ")}`);
    if (e.temporal?.valid_from) extra.push(`valid from: ${e.temporal.valid_from}`);
    if (e.temporal?.valid_until) extra.push(`valid until: ${e.temporal.valid_until}`);
    const suffix = extra.length > 0 ? ` (${extra.join("; ")})` : "";
    return `${i + 1}. [${e.id}] ${e.statement}${suffix}`;
  }).join("\n");
  const prompt = `You are a memory retrieval system. Given a query and a list of memories, select the ${limit} most relevant memories. Return ONLY the numbers of the relevant memories, comma-separated, in order of relevance (most relevant first).

Query: "${query}"

Memories:
${numbered}

Rules:
- Select at most ${limit} memories
- Only select memories that are actually relevant to answering the query
- If fewer than ${limit} memories are relevant, return fewer
- Return ONLY comma-separated numbers, nothing else (e.g., "3,7,1,12")
- If no memories are relevant, return "none"`;
  try {
    const response = await llm(prompt);
    const text = response.trim();
    if (text.toLowerCase() === "none") return [];
    const indices = text.split(",").map((s) => parseInt(s.trim(), 10) - 1).filter((i) => !isNaN(i) && i >= 0 && i < candidates.length);
    const seen = /* @__PURE__ */ new Set();
    const unique = [];
    for (const i of indices) {
      if (!seen.has(i)) {
        seen.add(i);
        unique.push(i);
      }
    }
    return unique.slice(0, limit).map((i) => candidates[i]);
  } catch {
    return candidates.slice(0, limit);
  }
}

// src/hybrid-search.ts
function rrfMerge(resultSets, k = 60) {
  const scores = /* @__PURE__ */ new Map();
  for (const results of resultSets) {
    for (let rank = 0; rank < results.length; rank++) {
      const engram = results[rank];
      const existing = scores.get(engram.id);
      const rrfScore = 1 / (k + rank + 1);
      if (existing) {
        existing.score += rrfScore;
      } else {
        scores.set(engram.id, { engram, score: rrfScore });
      }
    }
  }
  return Array.from(scores.values()).sort((a, b) => b.score - a.score).map((s) => s.engram);
}
var AGGREGATION_PATTERNS = [
  /how many/i,
  /how much/i,
  /total (?:number|amount|cost|time|hours|days|spent)/i,
  /all (?:the|my)/i,
  /every (?:time|instance)/i,
  /(?:count|sum|add up|combine|altogether)/i,
  /in the (?:past|last) (?:week|month|year|few)/i,
  /did I (?:attend|visit|go to|buy|spend|do)/i
];
function isAggregationQuery(query) {
  return AGGREGATION_PATTERNS.some((p) => p.test(query));
}
async function hybridSearch(engrams, query, limit, storagePath) {
  if (engrams.length === 0) return [];
  const exhaustive = isAggregationQuery(query);
  const effectiveLimit = exhaustive ? Math.max(limit, 50) : limit;
  const bm25Limit = Math.min(engrams.length, exhaustive ? effectiveLimit * 5 : effectiveLimit * 3);
  const embLimit = Math.min(engrams.length, exhaustive ? effectiveLimit * 3 : effectiveLimit * 2);
  const [bm25Results, embResults] = await Promise.all([
    Promise.resolve(searchEngrams(engrams, query, bm25Limit)),
    embeddingSearch(engrams, query, embLimit, storagePath)
  ]);
  const merged = rrfMerge([bm25Results, embResults]);
  return merged.slice(0, effectiveLimit);
}

// src/query-expansion.ts
var EXPANSION_PROMPT = `You are a search query expander. Given a user query, generate 3 alternative search queries that capture the same intent but use different words, synonyms, or related concepts.

Return ONLY the 3 queries, one per line, nothing else. No numbering, no explanations.

Query: "{query}"`;
var AGGREGATION_EXPANSION_PROMPT = `You are a search query expander for an aggregation query. The user wants to COUNT or TOTAL something across multiple conversations. Generate 5 search queries that will find ALL individual instances \u2014 each query should target a different way the item might have been mentioned.

For example, if the query is "How many weddings did I attend?", generate queries like:
- wedding ceremony attended
- went to wedding reception
- wedding invitation RSVP
- marriage celebration
- attended wedding of

Return ONLY the 5 queries, one per line. No numbering, no explanations.

Query: "{query}"`;
var AGGREGATION_PATTERNS2 = [
  /how many/i,
  /how much/i,
  /total (?:number|amount|cost|time|hours|days|spent)/i,
  /all (?:the|my)/i,
  /every (?:time|instance)/i,
  /(?:count|sum|add up|combine|altogether)/i,
  /in the (?:past|last) (?:week|month|year|few)/i,
  /did I (?:attend|visit|go to|buy|spend|do)/i
];
function isAggregationQuery2(query) {
  return AGGREGATION_PATTERNS2.some((p) => p.test(query));
}
function parseVariants(response, original) {
  const variants = response.split("\n").map((line) => line.replace(/^\d+[\.\)]\s*/, "").trim()).filter((line) => line.length > 3 && line.length < 200).slice(0, 3);
  return [original, ...variants];
}
function rrfMerge2(resultSets, k = 60) {
  const scores = /* @__PURE__ */ new Map();
  for (const results of resultSets) {
    for (let rank = 0; rank < results.length; rank++) {
      const engram = results[rank];
      const existing = scores.get(engram.id);
      const rrfScore = 1 / (k + rank + 1);
      if (existing) {
        existing.score += rrfScore;
      } else {
        scores.set(engram.id, { engram, score: rrfScore });
      }
    }
  }
  return Array.from(scores.values()).sort((a, b) => b.score - a.score).map((s) => s.engram);
}
async function expandedSearch(engrams, query, limit, llm, storagePath) {
  if (engrams.length === 0) return [];
  const aggregation = isAggregationQuery2(query);
  let variants;
  try {
    const promptTemplate = aggregation ? AGGREGATION_EXPANSION_PROMPT : EXPANSION_PROMPT;
    const prompt = promptTemplate.replace("{query}", query);
    const response = await llm(prompt);
    variants = parseVariants(response, query);
    if (aggregation) {
      const extra = response.split("\n").map((line) => line.replace(/^\d+[\.\)]\s*/, "").trim()).filter((line) => line.length > 3 && line.length < 200).slice(0, 5);
      variants = [query, ...extra];
    }
  } catch {
    variants = [query];
  }
  const perVariantLimit = aggregation ? Math.max(limit, 50) : limit;
  const searchPromises = variants.map(
    (v) => hybridSearch(engrams, v, perVariantLimit, storagePath)
  );
  const resultSets = await Promise.all(searchPromises);
  const merged = rrfMerge2(resultSets);
  const effectiveLimit = aggregation ? Math.max(limit, 50) : limit;
  return merged.slice(0, effectiveLimit);
}

// src/packs.ts
import * as fs2 from "fs";
import * as path from "path";
import * as crypto from "crypto";
import yaml4 from "js-yaml";

// src/secrets.ts
var SECRET_PATTERNS = [
  { name: "aws_access_key", regex: /AKIA[0-9A-Z]{16}/ },
  { name: "aws_secret_key", regex: /(?:aws_secret_access_key|secret_access_key)\s*[=:]\s*[A-Za-z0-9/+=]{40}/i },
  { name: "generic_api_key", regex: /(?:^|[^a-z])(sk|pk)[-_][a-z0-9]{20,}/i },
  { name: "api_key_assignment", regex: /(?:api[_-]?key|api[_-]?secret|secret[_-]?key)\s*[=:]\s*\S{20,}/i },
  { name: "password_assignment", regex: /password\s*[=:]\s*\S{8,}/i },
  { name: "connection_string", regex: /(?:postgres|mysql|mongodb|redis):\/\/\S+/ },
  { name: "jwt", regex: /eyJ[A-Za-z0-9_-]{10,}\.eyJ[A-Za-z0-9_-]{10,}/ },
  { name: "private_key", regex: /-----BEGIN\s+\S+\s+PRIVATE KEY-----/ },
  { name: "bearer_token", regex: /Bearer\s+[A-Za-z0-9._~+/=-]{20,}/ }
];
function detectSecrets(text) {
  const matches = [];
  for (const { name, regex } of SECRET_PATTERNS) {
    const m = text.match(regex);
    if (m) {
      matches.push({ pattern: name, match: m[0].slice(0, 20) + "..." });
    }
  }
  return matches;
}

// src/packs.ts
function detectConflicts2(newEngrams, existingEngrams) {
  const conflicts = [];
  for (const ne of newEngrams) {
    for (const ee of existingEngrams) {
      const nNorm = ne.statement.toLowerCase().replace(/\s+/g, " ").trim();
      const eNorm = ee.statement.toLowerCase().replace(/\s+/g, " ").trim();
      if (nNorm === eNorm) {
        conflicts.push({
          pack_engram_id: ne.id,
          pack_statement: ne.statement.slice(0, 120),
          existing_engram_id: ee.id,
          existing_statement: ee.statement.slice(0, 120),
          type: "duplicate"
        });
        continue;
      }
      if (ne.domain && ee.domain && ne.domain === ee.domain) {
        const nHasNever = /\b(never|don't|do not|avoid|stop)\b/i.test(ne.statement);
        const eHasNever = /\b(never|don't|do not|avoid|stop)\b/i.test(ee.statement);
        const nHasAlways = /\b(always|must|should|prefer|use)\b/i.test(ne.statement);
        const eHasAlways = /\b(always|must|should|prefer|use)\b/i.test(ee.statement);
        if (nHasNever && eHasAlways || nHasAlways && eHasNever) {
          const nWords = new Set(nNorm.split(" ").filter((w) => w.length > 4));
          const eWords = new Set(eNorm.split(" ").filter((w) => w.length > 4));
          const overlap = [...nWords].filter((w) => eWords.has(w));
          if (overlap.length >= 2) {
            conflicts.push({
              pack_engram_id: ne.id,
              pack_statement: ne.statement.slice(0, 120),
              existing_engram_id: ee.id,
              existing_statement: ee.statement.slice(0, 120),
              type: "contradiction"
            });
          }
        }
      }
    }
  }
  return conflicts;
}
function installPack(packsDir, source, existingEngrams) {
  if (!fs2.existsSync(source)) throw new Error(`Pack source not found: ${source}`);
  const sourceName = path.basename(source);
  const destDir = path.join(packsDir, sourceName);
  if (!fs2.existsSync(destDir)) fs2.mkdirSync(destDir, { recursive: true });
  const files = fs2.readdirSync(source);
  for (const file of files) {
    const srcPath = path.join(source, file);
    const destPath = path.join(destDir, file);
    if (fs2.statSync(srcPath).isFile()) {
      fs2.copyFileSync(srcPath, destPath);
    }
  }
  const engramsPath = path.join(destDir, "engrams.yaml");
  const newEngrams = fs2.existsSync(engramsPath) ? loadEngrams(engramsPath) : [];
  const conflicts = existingEngrams ? detectConflicts2(newEngrams, existingEngrams) : [];
  return { installed: newEngrams.length, name: sourceName, conflicts };
}
function uninstallPack(packsDir, name) {
  let packDir = path.join(packsDir, name);
  if (!fs2.existsSync(packDir)) {
    const entries = fs2.existsSync(packsDir) ? fs2.readdirSync(packsDir) : [];
    const match = entries.find((e) => e.toLowerCase() === name.toLowerCase());
    if (match) {
      packDir = path.join(packsDir, match);
    } else {
      throw new Error(`Pack not found: ${name}. Use 'plur packs list' to see installed packs.`);
    }
  }
  const engramsPath = path.join(packDir, "engrams.yaml");
  let count = 0;
  try {
    count = loadEngrams(engramsPath).length;
  } catch {
  }
  fs2.rmSync(packDir, { recursive: true, force: true });
  return { name, removed: true, engram_count: count };
}
function listPacks(packsDir) {
  if (!fs2.existsSync(packsDir)) return [];
  const result = [];
  for (const entry of fs2.readdirSync(packsDir)) {
    const packDir = path.join(packsDir, entry);
    if (!fs2.statSync(packDir).isDirectory()) continue;
    try {
      const pack = loadPack(packDir);
      result.push({
        name: pack.manifest.name,
        path: packDir,
        engram_count: pack.engrams.length,
        manifest: pack.manifest,
        integrity: computePackHash(packDir)
      });
    } catch {
      const engramsPath = path.join(packDir, "engrams.yaml");
      const engrams = fs2.existsSync(engramsPath) ? loadEngrams(engramsPath) : [];
      result.push({
        name: entry,
        path: packDir,
        engram_count: engrams.length
      });
    }
  }
  return result;
}
var PERSONAL_PATH_RE = /(?:\/Users\/\w+|\/home\/\w+|~\/|C:\\Users\\\w+)/;
var EMAIL_RE = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/;
var IP_RE = /\b(?:10|172\.(?:1[6-9]|2\d|3[01])|192\.168)\.\d{1,3}\.\d{1,3}\b/;
function scanPrivacy(engrams) {
  const issues = [];
  for (const e of engrams) {
    if (e.visibility === "private") {
      issues.push({
        engram_id: e.id,
        type: "private_visibility",
        detail: `Engram marked as private \u2014 skipped from export`
      });
      continue;
    }
    const text = `${e.statement} ${e.rationale ?? ""} ${e.source ?? ""}`;
    const secrets = detectSecrets(text);
    for (const s of secrets) {
      issues.push({
        engram_id: e.id,
        type: "secret",
        detail: `${s.pattern}: ${s.match}`
      });
    }
    if (PERSONAL_PATH_RE.test(text)) {
      issues.push({
        engram_id: e.id,
        type: "personal_path",
        detail: `Contains personal path: ${text.match(PERSONAL_PATH_RE)?.[0]}`
      });
    }
    const emailMatch = text.match(EMAIL_RE);
    if (emailMatch) {
      issues.push({
        engram_id: e.id,
        type: "email",
        detail: `Contains email: ${emailMatch[0]}`
      });
    }
    const ipMatch = text.match(IP_RE);
    if (ipMatch) {
      issues.push({
        engram_id: e.id,
        type: "ip_address",
        detail: `Contains private IP: ${ipMatch[0]}`
      });
    }
  }
  return { clean: issues.length === 0, issues };
}
function deriveMatchTerms(engrams) {
  const termCounts = /* @__PURE__ */ new Map();
  for (const e of engrams) {
    if (e.tags) {
      for (const t of e.tags) {
        termCounts.set(t, (termCounts.get(t) || 0) + 1);
      }
    }
    if (e.domain) {
      for (const part of e.domain.split(".")) {
        if (part.length > 2) {
          termCounts.set(part, (termCounts.get(part) || 0) + 1);
        }
      }
    }
    if (e.type) {
      termCounts.set(e.type, (termCounts.get(e.type) || 0) + 1);
    }
  }
  return [...termCounts.entries()].filter(([, count]) => count >= 2).sort((a, b) => b[1] - a[1]).slice(0, 20).map(([term]) => term);
}
function exportPack(engrams, outputDir, manifest) {
  const allPrivacy = scanPrivacy(engrams);
  const blockedIds = new Set(
    allPrivacy.issues.filter((i) => i.type === "secret" || i.type === "private_visibility").map((i) => i.engram_id)
  );
  const safeEngrams = engrams.filter((e) => !blockedIds.has(e.id));
  const matchTerms = deriveMatchTerms(safeEngrams);
  if (!fs2.existsSync(outputDir)) fs2.mkdirSync(outputDir, { recursive: true });
  const frontmatter = yaml4.dump({
    name: manifest.name,
    version: manifest.version,
    description: manifest.description,
    creator: manifest.creator,
    metadata: {
      injection_policy: "on_match",
      match_terms: matchTerms,
      engram_count: safeEngrams.length
    }
  });
  fs2.writeFileSync(
    path.join(outputDir, "SKILL.md"),
    `---
${frontmatter}---

# ${manifest.name}

${manifest.description || ""}
`
  );
  const exportEngrams = safeEngrams.map((e) => {
    const cleaned = { ...e };
    if (cleaned.relations) {
      cleaned.relations = {
        ...cleaned.relations,
        conflicts: [],
        related: []
      };
    }
    if (cleaned.associations) {
      cleaned.associations = [];
    }
    if (cleaned.knowledge_anchors) {
      cleaned.knowledge_anchors = [];
    }
    if (cleaned.activation) {
      cleaned.activation = {
        ...cleaned.activation,
        frequency: 0,
        retrieval_strength: 0.7
      };
    }
    if (cleaned.feedback_signals) {
      cleaned.feedback_signals = { positive: 0, negative: 0, neutral: 0 };
    }
    return cleaned;
  });
  const content = yaml4.dump({ engrams: exportEngrams }, { lineWidth: 120, noRefs: true, quotingType: '"' });
  fs2.writeFileSync(path.join(outputDir, "engrams.yaml"), content);
  const integrity = computePackHash(outputDir);
  fs2.writeFileSync(path.join(outputDir, "INTEGRITY"), `sha256:${integrity}
`);
  return {
    path: outputDir,
    engram_count: safeEngrams.length,
    privacy: allPrivacy,
    match_terms: matchTerms,
    integrity: `sha256:${integrity}`
  };
}
function computePackHash(packDir) {
  const hash = crypto.createHash("sha256");
  const skillMd = path.join(packDir, "SKILL.md");
  const manifestYaml = path.join(packDir, "manifest.yaml");
  if (fs2.existsSync(skillMd)) {
    hash.update(fs2.readFileSync(skillMd));
  } else if (fs2.existsSync(manifestYaml)) {
    hash.update(fs2.readFileSync(manifestYaml));
  }
  const engramsPath = path.join(packDir, "engrams.yaml");
  if (fs2.existsSync(engramsPath)) {
    hash.update(fs2.readFileSync(engramsPath));
  }
  return hash.digest("hex");
}

// src/meta/sanitize.ts
function sanitizeForPrompt(text) {
  return text.replace(/```/g, "~~~").replace(/\n{3,}/g, "\n\n").replace(/^(system|assistant|user):/gim, "$1 -").slice(0, 2e3);
}

// src/meta/structural-analysis.ts
var FAILURE_TAGS = /* @__PURE__ */ new Set(["correction", "mistake", "bug", "fix", "error", "wrong", "broken"]);
var BATCH_SIZE = 5;
function isFailureDriven(engram) {
  if (engram.tags.some((t) => FAILURE_TAGS.has(t.toLowerCase()))) return true;
  const fb = engram.feedback_signals;
  if (fb && fb.negative > fb.positive) return true;
  return false;
}
function prioritize(engrams) {
  return [...engrams].sort((a, b) => {
    const aFail = isFailureDriven(a) ? 0 : 1;
    const bFail = isFailureDriven(b) ? 0 : 1;
    if (aFail !== bFail) return aFail - bFail;
    const aFb = a.feedback_signals?.negative ?? 0;
    const bFb = b.feedback_signals?.negative ?? 0;
    return bFb - aFb;
  });
}
function buildPrompt(engrams) {
  const items = engrams.map((e, i) => {
    const parts = [`  Statement: "${sanitizeForPrompt(e.statement)}"`];
    if (e.rationale) parts.push(`  Rationale: "${sanitizeForPrompt(e.rationale)}"`);
    if (e.domain) parts.push(`  Domain: "${e.domain}"`);
    return `Engram ${i + 1} (${e.id}):
${parts.join("\n")}`;
  }).join("\n\n");
  return `Given these learned lessons, extract relational structure for cross-domain transfer.

${items}

For EACH engram, extract:
1. The GOAL the agent had (abstract type, not domain-specific)
2. Relational triples: subject (role + domain instance), predicate, object (role + domain instance), outcome
3. Use domain-GENERAL role types (e.g., "safety-metric" not "liquidation-price")

Return a JSON array with one object per engram:
[{
  "engram_id": "ENG-...",
  "goal_context": "what the agent was trying to achieve (abstract)",
  "triples": [{ "subject": { "role": "...", "domain_instance": "..." }, "predicate": "...", "object": { "role": "...", "domain_instance": "..." }, "outcome": "..." }]
}]

Rules:
- Roles MUST be domain-general (e.g., "safety-metric" not "liquidation-price")
- If a lesson is purely domain-specific with no transferable structure, return empty triples: []
- Prefer causal chains over isolated attributes
- Return ONLY valid JSON, no markdown fencing`;
}
async function analyzeStructure(engrams, llm) {
  const prioritized = prioritize(engrams.filter((e) => e.status === "active"));
  const results = [];
  for (let i = 0; i < prioritized.length; i += BATCH_SIZE) {
    const batch = prioritized.slice(i, i + BATCH_SIZE);
    const prompt = buildPrompt(batch);
    let parsed;
    try {
      const response = await llm(prompt);
      const cleaned = response.replace(/```json\n?/g, "").replace(/```\n?/g, "").trim();
      parsed = JSON.parse(cleaned);
    } catch {
      try {
        const retryResponse = await llm(prompt + "\n\nIMPORTANT: Return ONLY raw JSON array, no text.");
        const cleaned = retryResponse.replace(/```json\n?/g, "").replace(/```\n?/g, "").trim();
        parsed = JSON.parse(cleaned);
      } catch {
        continue;
      }
    }
    for (const item of parsed) {
      if (!item.triples || item.triples.length === 0) continue;
      const engram = batch.find((e) => e.id === item.engram_id) ?? batch[parsed.indexOf(item)];
      if (!engram) continue;
      results.push({
        engram_id: engram.id,
        triples: item.triples,
        goal_context: item.goal_context ?? "",
        is_failure_driven: isFailureDriven(engram),
        domain: engram.domain ?? "unknown",
        polarity: classifyPolarity(engram.statement) === "dont" ? "dont" : "do"
      });
    }
  }
  return results;
}

// src/meta/similarity.ts
function tokenSimilarity(a, b) {
  const wordsA = new Set(a.toLowerCase().split(/[\s\[\]→+\-]+/).filter((w) => w.length > 2));
  const wordsB = new Set(b.toLowerCase().split(/[\s\[\]→+\-]+/).filter((w) => w.length > 2));
  if (wordsA.size === 0 || wordsB.size === 0) return 0;
  let overlap = 0;
  for (const w of wordsA) {
    if (wordsB.has(w)) overlap++;
  }
  return overlap / Math.sqrt(wordsA.size * wordsB.size);
}

// src/meta/clustering.ts
function tripleToTemplate(triple) {
  const parts = [triple.subject.role, triple.predicate, triple.object.role];
  if (triple.outcome) parts.push("\u2192", triple.outcome);
  return parts.join(" ");
}
function analysisTemplate(analysis) {
  return analysis.triples.map(tripleToTemplate).join("; ");
}
async function computeSimilarityMatrix(templates) {
  const n = templates.length;
  const matrix = Array.from({ length: n }, () => Array(n).fill(0));
  try {
    const { embed, cosineSimilarity } = await import("./embeddings-2IODIQAF.js");
    const embeddings = [];
    for (const t of templates) {
      embeddings.push(await embed(t));
    }
    if (embeddings[0] !== null) {
      for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
          if (embeddings[i] && embeddings[j]) {
            const sim = cosineSimilarity(embeddings[i], embeddings[j]);
            matrix[i][j] = sim;
            matrix[j][i] = sim;
          }
        }
      }
      return { matrix, method: "embedding" };
    }
  } catch {
  }
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const sim = tokenSimilarity(templates[i], templates[j]);
      matrix[i][j] = sim;
      matrix[j][i] = sim;
    }
  }
  return { matrix, method: "token" };
}
var EMBEDDING_THRESHOLD = 0.65;
var TOKEN_THRESHOLD = 0.35;
async function clusterByStructure(analyses, threshold2) {
  if (analyses.length < 2) return [];
  const templates = analyses.map((a) => analysisTemplate(a));
  const { matrix, method } = await computeSimilarityMatrix(templates);
  const effectiveThreshold = threshold2 ?? (method === "embedding" ? EMBEDDING_THRESHOLD : TOKEN_THRESHOLD);
  const seenClusters = /* @__PURE__ */ new Set();
  const clusters = [];
  let clusterId = 0;
  for (let i = 0; i < analyses.length; i++) {
    const members = [analyses[i]];
    const memberIndices = [i];
    for (let j = 0; j < analyses.length; j++) {
      if (j === i) continue;
      if (matrix[i][j] >= effectiveThreshold) {
        members.push(analyses[j]);
        memberIndices.push(j);
      }
    }
    if (members.length < 2) continue;
    const memberKey = members.map((m) => m.engram_id).sort().join(",");
    if (seenClusters.has(memberKey)) continue;
    seenClusters.add(memberKey);
    const domains = [...new Set(members.map((m) => m.domain))];
    let totalSim = 0;
    let pairs = 0;
    for (let a = 0; a < memberIndices.length; a++) {
      for (let b = a + 1; b < memberIndices.length; b++) {
        totalSim += matrix[memberIndices[a]][memberIndices[b]];
        pairs++;
      }
    }
    clusters.push({
      cluster_id: `cluster-${clusterId++}`,
      members,
      domains,
      is_cross_domain: domains.length >= 2,
      cohesion: pairs > 0 ? totalSim / pairs : 0
    });
  }
  return clusters;
}

// src/meta/platitudes.ts
var PLATITUDE_PATTERNS = [
  "always be careful",
  "verify before acting",
  "plan ahead",
  "consider edge cases",
  "communicate clearly",
  "test thoroughly",
  "keep it simple",
  "document your work",
  "think before you act",
  "double check"
];
function isPlatitude(statement) {
  const lower = statement.toLowerCase();
  return PLATITUDE_PATTERNS.some((p) => lower.includes(p)) || statement.length < 30;
}

// src/meta/alignment.ts
async function alignCluster(cluster, llm) {
  const tripleDescriptions = cluster.members.map((m) => {
    const domain = sanitizeForPrompt(m.domain);
    const triples = m.triples.map((t) => {
      const parts = [`${sanitizeForPrompt(t.subject.role)} ${sanitizeForPrompt(t.predicate)} ${sanitizeForPrompt(t.object.role)}`];
      if (t.outcome) parts.push(`\u2192 ${sanitizeForPrompt(t.outcome)}`);
      return parts.join(" ");
    }).join("; ");
    return `Domain ${domain} (${m.engram_id}): ${triples}`;
  }).join("\n");
  const prompt = `These lessons come from different domains but may share a common structural principle:

${tripleDescriptions}

Tasks:
1. Is there a genuine common relational structure? (Not surface similarity, not a platitude)
2. Choose the structural frame that BEST captures the pattern:
   - "goal-constraint-outcome": [goal] + [constraint] \u2192 [outcome]
   - "feedback-loop": [action] \u2192 [effect] \u2192 [feeds back to action]
   - "causal-chain": [A] causes [B] causes [C]
   - "tradeoff": [optimizing X] at the expense of [Y]
   - "recursive": [pattern] contains smaller instances of [same pattern]
   - "freeform": describe the structure in your own terms
3. Express the common structure using your chosen frame
4. Rate structural depth (1-5): 1=surface, 3=causal chain, 5=deep systematic
5. For each member, rate alignment (0-1) and explain the mapping in one sentence
6. What does this structure PREDICT in a domain not listed above?

Return JSON:
{
  "common_structure": {
    "goal_type": "...",
    "constraint_type": "...",
    "outcome_type": "...",
    "template": "...",
    "structure_type": "goal-constraint-outcome" | "feedback-loop" | "causal-chain" | "tradeoff" | "recursive" | "freeform",
    "freeform_structure": "..." (only if structure_type is "freeform")
  } | null,
  "structural_depth": N,
  "member_alignments": [{ "engram_id": "...", "alignment_score": N, "mapping_rationale": "..." }],
  "candidate_inferences": ["..."]
}

If the commonality is shallow or trivially obvious (e.g., "mistakes happen", "always be careful"), return null for common_structure. We prefer no extraction over a false abstraction.
Return ONLY valid JSON, no markdown fencing.`;
  try {
    const response = await llm(prompt);
    const cleaned = response.replace(/```json\n?/g, "").replace(/```\n?/g, "").trim();
    if (cleaned === "null" || cleaned === '{"common_structure": null}') return null;
    const parsed = JSON.parse(cleaned);
    if (!parsed.common_structure) return null;
    if (parsed.structural_depth < 2) return null;
    if (isPlatitude(parsed.common_structure.template)) return null;
    if (parsed.common_structure.template.length < 30) return null;
    const alignments = parsed.member_alignments ?? [];
    const lowAlignments = alignments.filter((a) => a.alignment_score < 0.5);
    if (lowAlignments.length > alignments.length / 2) return null;
    return {
      cluster_id: cluster.cluster_id,
      common_structure: parsed.common_structure,
      member_alignments: alignments,
      structural_depth: parsed.structural_depth,
      systematicity: parsed.structural_depth,
      // Use depth as systematicity proxy
      candidate_inferences: parsed.candidate_inferences ?? []
    };
  } catch {
    return null;
  }
}

// src/meta/formulation.ts
var PIPELINE_VERSION = "1.0.0";
function slugifyTemplate(template) {
  return template.toLowerCase().replace(/[\[\]→+]/g, "").replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 60);
}
function isDuplicate(template, existingMetas) {
  for (const meta of existingMetas) {
    const metaField = meta.structured_data?.meta;
    if (!metaField?.structure?.template) continue;
    const sim = tokenSimilarity(template, metaField.structure.template);
    if (sim > 0.9) return meta;
  }
  return null;
}
async function formulateMetaEngram(alignment, llm, existingMetas = [], domains = []) {
  const { common_structure, member_alignments, structural_depth, candidate_inferences } = alignment;
  if (isPlatitude(common_structure.template)) return null;
  const memberSummary = member_alignments.map((m) => `  - ${m.engram_id}: ${sanitizeForPrompt(m.mapping_rationale)} (alignment: ${m.alignment_score})`).join("\n");
  const prompt = `Structural principle: ${sanitizeForPrompt(common_structure.template)}
Goal type: ${common_structure.goal_type}
Constraint type: ${common_structure.constraint_type}
Outcome type: ${common_structure.outcome_type}
Structural depth: ${structural_depth}
Evidence from members:
${memberSummary}
Candidate inferences: ${candidate_inferences.map(sanitizeForPrompt).join(", ")}

Generate:
1. A natural-language statement of this principle (1-2 sentences, precise, not generic)
2. Conditions under which it holds (expected_conditions)
3. Conditions under which it does NOT hold (expected_exceptions)
4. A concrete test: pick a domain NOT in the evidence list, describe a scenario where this principle predicts a specific outcome (test_prediction)

The statement must be specific enough to be FALSIFIABLE. If you cannot write falsification criteria, return null.

Return JSON:
{
  "statement": "...",
  "falsification": {
    "expected_conditions": "...",
    "expected_exceptions": "...",
    "test_prediction": "..."
  }
}

Or return null if the principle is too vague to falsify.
Return ONLY valid JSON, no markdown fencing.`;
  let parsed;
  try {
    const response = await llm(prompt);
    const cleaned = response.replace(/```json\n?/g, "").replace(/```\n?/g, "").trim();
    if (cleaned === "null") return null;
    parsed = JSON.parse(cleaned);
  } catch {
    return null;
  }
  if (!parsed || !parsed.statement || !parsed.falsification) return null;
  if (!parsed.falsification.expected_conditions || !parsed.falsification.expected_exceptions) return null;
  const duplicate = isDuplicate(common_structure.template, existingMetas);
  if (duplicate) {
    return null;
  }
  const evidenceCount = member_alignments.length;
  const domainCount = domains.length > 0 ? domains.length : 1;
  const validationRatio = 0;
  const composite = computeMetaConfidence(evidenceCount, domainCount, structural_depth, validationRatio);
  const slug = slugifyTemplate(common_structure.template);
  const id = `META-${slug}`;
  const evidence = member_alignments.map((ma) => ({
    engram_id: ma.engram_id,
    domain: "unknown",
    // Will be enriched by pipeline orchestrator
    mapping_rationale: ma.mapping_rationale,
    alignment_score: ma.alignment_score
  }));
  const metaField = {
    structure: common_structure,
    evidence,
    domain_coverage: {
      validated: [],
      failed: [],
      predicted: candidate_inferences
    },
    falsification: parsed.falsification,
    confidence: {
      evidence_count: evidenceCount,
      domain_count: domainCount,
      structural_depth,
      validation_ratio: validationRatio,
      composite
    },
    hierarchy: {
      level: domainCount >= 3 ? "top" : "mop",
      parent: null,
      children: []
    },
    pipeline_version: PIPELINE_VERSION
  };
  const now = (/* @__PURE__ */ new Date()).toISOString();
  const engram = {
    id,
    version: 2,
    status: "active",
    consolidated: false,
    type: "behavioral",
    scope: "global",
    visibility: "private",
    statement: parsed.statement,
    domain: "meta",
    tags: ["meta-engram", common_structure.goal_type, common_structure.constraint_type],
    activation: {
      retrieval_strength: composite,
      storage_strength: 1,
      frequency: 0,
      last_accessed: now.slice(0, 10)
    },
    feedback_signals: { positive: 0, negative: 0, neutral: 0 },
    knowledge_anchors: [],
    associations: [],
    derivation_count: evidenceCount,
    pack: null,
    abstract: null,
    derived_from: null,
    polarity: null,
    knowledge_type: {
      memory_class: "metacognitive",
      cognitive_level: "evaluate"
    },
    structured_data: { meta: metaField }
  };
  return engram;
}

// src/meta/validation.ts
async function validateMetaEngram(meta, testEngrams, testDomain, llm) {
  const metaField = meta.structured_data?.meta;
  if (!metaField) {
    return {
      meta_engram_id: meta.id,
      test_domain: testDomain,
      prediction_held: false,
      matching_engram_id: null,
      alignment_score: 0,
      rationale: "Meta-engram has no meta field"
    };
  }
  const template = metaField.structure.template;
  const testEngramSummary = testEngrams.filter((e) => e.status === "active").slice(0, 10).map((e) => `  - [${e.id}] ${sanitizeForPrompt(e.statement)}`).join("\n");
  if (!testEngramSummary) {
    return {
      meta_engram_id: meta.id,
      test_domain: testDomain,
      prediction_held: false,
      matching_engram_id: null,
      alignment_score: 0,
      rationale: "No test engrams available for validation"
    };
  }
  const prompt = `Meta-engram structural principle: ${sanitizeForPrompt(template)}
Meta-engram statement: ${sanitizeForPrompt(meta.statement)}

Test domain: ${testDomain}
Test engrams from this domain:
${testEngramSummary}

Task: Does any of the test engrams instantiate or validate this structural principle?

If YES (a test engram matches):
- prediction_held: true
- matching_engram_id: the ID of the matching engram
- alignment_score: 0.5-1.0 (how well the structural template maps)
- rationale: one sentence explaining how the test engram maps to the template

If NO (no test engram matches):
- prediction_held: false
- matching_engram_id: null
- alignment_score: 0
- rationale: one sentence explaining WHY none of the test engrams match (e.g., "Test engrams cover X domain but none exhibit the structural pattern of [assumed independence] \u2192 [understated risk]")

Return JSON:
{
  "prediction_held": true|false,
  "matching_engram_id": "ENG-..." | null,
  "alignment_score": 0.0-1.0,
  "rationale": "one sentence explanation \u2014 REQUIRED even when prediction_held is false"
}

Return ONLY valid JSON, no markdown fencing.`;
  try {
    const response = await llm(prompt);
    const cleaned = response.replace(/```json\n?/g, "").replace(/```\n?/g, "").trim();
    const parsed = JSON.parse(cleaned);
    const result = {
      meta_engram_id: meta.id,
      test_domain: testDomain,
      prediction_held: Boolean(parsed.prediction_held),
      matching_engram_id: parsed.matching_engram_id ?? null,
      alignment_score: typeof parsed.alignment_score === "number" ? parsed.alignment_score : 0,
      rationale: parsed.rationale ?? ""
    };
    if (result.prediction_held) {
      if (!metaField.domain_coverage.validated.includes(testDomain)) {
        metaField.domain_coverage.validated.push(testDomain);
      }
    } else {
      if (!metaField.domain_coverage.failed.includes(testDomain)) {
        metaField.domain_coverage.failed.push(testDomain);
      }
    }
    const passes = metaField.domain_coverage.validated.length;
    const total = passes + metaField.domain_coverage.failed.length;
    const validationRatio = total > 0 ? passes / total : 0;
    const updatedComposite = computeMetaConfidence(
      metaField.confidence.evidence_count,
      metaField.confidence.domain_count,
      metaField.confidence.structural_depth,
      validationRatio
    );
    metaField.confidence.validation_ratio = validationRatio;
    metaField.confidence.composite = updatedComposite;
    metaField.last_validated = (/* @__PURE__ */ new Date()).toISOString().slice(0, 10);
    const failedCount = metaField.domain_coverage.failed.length;
    if (failedCount >= 3) {
      if (metaField.hierarchy.level === "top") {
        metaField.hierarchy.level = "mop";
      } else {
        meta.status = "retired";
      }
    }
    return result;
  } catch {
    return {
      meta_engram_id: meta.id,
      test_domain: testDomain,
      prediction_held: false,
      matching_engram_id: null,
      alignment_score: 0,
      rationale: "LLM parsing failed during validation"
    };
  }
}

// src/meta/hierarchy.ts
var SUBSUMPTION_THRESHOLD = 0.75;
function getMetaField(engram) {
  return engram.structured_data?.meta ?? null;
}
function getDomainCount(metaField) {
  const validated = metaField.domain_coverage.validated.length;
  const evidenceDomains = new Set(metaField.evidence.map((e) => e.domain)).size;
  return Math.max(validated, evidenceDomains, metaField.confidence.domain_count);
}
function organizeHierarchy(metaEngrams) {
  const metas = metaEngrams.filter((e) => getMetaField(e) !== null && e.id.startsWith("META-"));
  if (metas.length === 0) return metaEngrams;
  for (const meta of metas) {
    const mf = getMetaField(meta);
    mf.hierarchy.parent = null;
    mf.hierarchy.children = [];
  }
  for (let i = 0; i < metas.length; i++) {
    for (let j = 0; j < metas.length; j++) {
      if (i === j) continue;
      const mfI = getMetaField(metas[i]);
      const mfJ = getMetaField(metas[j]);
      const sim = tokenSimilarity(mfI.structure.template, mfJ.structure.template);
      if (sim < SUBSUMPTION_THRESHOLD) continue;
      const domainsI = getDomainCount(mfI);
      const domainsJ = getDomainCount(mfJ);
      if (domainsI > domainsJ) {
        if (!mfI.hierarchy.children.includes(metas[j].id)) {
          mfI.hierarchy.children.push(metas[j].id);
        }
        if (!mfJ.hierarchy.parent) {
          mfJ.hierarchy.parent = metas[i].id;
        } else {
          const currentParent = metas.find((m) => m.id === mfJ.hierarchy.parent);
          const currentParentMf = currentParent ? getMetaField(currentParent) : null;
          if (currentParentMf && getDomainCount(currentParentMf) < domainsI) {
            mfJ.hierarchy.parent = metas[i].id;
          }
        }
      }
    }
  }
  for (const meta of metas) {
    const mf = getMetaField(meta);
    const domains = getDomainCount(mf);
    if (domains >= 3 && !mf.hierarchy.parent) {
      mf.hierarchy.level = "top";
    } else {
      mf.hierarchy.level = "mop";
    }
  }
  return metaEngrams;
}

// src/meta/pipeline.ts
async function extractMetaEngrams(engrams, llm, options = {}) {
  const start = Date.now();
  const {
    run_validation = false,
    validation_engrams = [],
    existing_metas = [],
    min_cluster_size = 2
  } = options;
  let rejected_as_platitudes = 0;
  const analyses = await analyzeStructure(engrams, llm);
  const clusters = await clusterByStructure(analyses);
  const viableClusters = clusters.filter((c) => c.members.length >= min_cluster_size);
  const alignmentResults = await Promise.all(
    viableClusters.map((cluster) => alignCluster(cluster, llm))
  );
  const validAlignments = alignmentResults.filter((r) => r !== null);
  const formulated = [];
  for (const alignment of validAlignments) {
    const cluster = viableClusters.find((c) => c.cluster_id === alignment.cluster_id);
    const clusterDomains = cluster ? [...new Set(cluster.members.map((m) => m.domain))] : [];
    const meta = await formulateMetaEngram(alignment, llm, [...existing_metas, ...formulated], clusterDomains);
    if (meta === null) {
      rejected_as_platitudes++;
    } else {
      if (cluster) {
        const metaField = meta.structured_data?.meta;
        if (metaField) {
          const domainMap = new Map(cluster.members.map((m) => [m.engram_id, m.domain]));
          for (const ev of metaField.evidence) {
            if (domainMap.has(ev.engram_id)) {
              ev.domain = domainMap.get(ev.engram_id);
            }
          }
          metaField.domain_coverage.validated = clusterDomains;
          meta.domain = clusterDomains.length >= 3 ? "meta" : `meta.${clusterDomains[0] ?? "unknown"}`;
        }
      }
      formulated.push(meta);
    }
  }
  const validationResults = [];
  if (run_validation && validation_engrams.length > 0) {
    for (const meta of formulated) {
      const metaField = meta.structured_data?.meta;
      if (!metaField) continue;
      const evidenceDomains = new Set(metaField.evidence.map((e) => e.domain));
      const testEngrams = validation_engrams.filter((e) => !evidenceDomains.has(e.domain ?? ""));
      if (testEngrams.length > 0) {
        const byDomain = /* @__PURE__ */ new Map();
        for (const te of testEngrams) {
          const d = te.domain ?? "unknown";
          if (!byDomain.has(d)) byDomain.set(d, []);
          byDomain.get(d).push(te);
        }
        for (const [domain, domainEngrams] of byDomain) {
          const vr = await validateMetaEngram(meta, domainEngrams, domain, llm);
          validationResults.push(vr);
        }
      }
    }
  }
  const organized = organizeHierarchy(formulated);
  const duration_ms = Date.now() - start;
  return {
    engrams_analyzed: analyses.length,
    clusters_found: clusters.length,
    alignments_passed: validAlignments.length,
    meta_engrams_extracted: organized.length,
    rejected_as_platitudes,
    validation_results: validationResults,
    results: organized,
    duration_ms
  };
}

// src/session-state.ts
var SessionBreadcrumbs = class {
  toolCalls = [];
  engramsRecalled = [];
  recordToolCall(tool, args) {
    this.toolCalls.push({
      tool,
      args_keys: Object.keys(args).join(", "),
      timestamp: (/* @__PURE__ */ new Date()).toISOString()
    });
  }
  recordEngramRecalled(engramId) {
    if (!this.engramsRecalled.includes(engramId)) {
      this.engramsRecalled.push(engramId);
    }
  }
  getToolCalls() {
    return this.toolCalls;
  }
  getEngramsRecalled() {
    return this.engramsRecalled;
  }
  /**
   * Returns only engram IDs that have the META- prefix.
   * These represent meta-engrams injected into the session context.
   */
  getMetaEngramsRecalled() {
    return this.engramsRecalled.filter((id) => id.startsWith("META-"));
  }
  generateContinuationContext() {
    if (this.toolCalls.length === 0 && this.engramsRecalled.length === 0) return "";
    const lines = [];
    if (this.toolCalls.length > 0) {
      const uniqueTools = [...new Set(this.toolCalls.map((tc) => tc.tool))];
      lines.push(`Tools used: ${uniqueTools.join(", ")}`);
      lines.push(`Total tool calls: ${this.toolCalls.length}`);
    }
    if (this.engramsRecalled.length > 0) {
      lines.push(`Engrams recalled: ${this.engramsRecalled.length}`);
      const metaCount = this.getMetaEngramsRecalled().length;
      if (metaCount > 0) {
        lines.push(`Meta-engrams: ${metaCount}`);
      }
    }
    return lines.join("\n");
  }
};

// src/guardrails.ts
function generateGuardrails() {
  return `## PLUR Memory Guardrails

### Verification Protocol
When recalling facts that will drive actions (server IPs, file paths, API endpoints, credential locations):
1. State the recalled fact explicitly before acting on it
2. Include the engram ID or search that produced it
3. If no engram matches, say "No engram found \u2014 verifying from filesystem" and check directly
4. Never interpolate between two engrams to produce a "probably correct" composite

When the user corrects a recalled fact: call plur.learn immediately, then plur.feedback with negative signal on the wrong engram, before continuing the task.

### Over-engineering Check
Before proposing any new system, module, or architectural change:
1. What is the simplest version that solves the actual problem?
2. Is there an existing tool/pattern that already covers 80% of this?
3. Will this create maintenance burden disproportionate to its value?

If a task can be done in <20 lines of shell script, do that first.

### Tool Selection Discipline
Before invoking any external tool, apply the locality test:
1. Is the answer already in engrams? \u2192 plur.recall
2. Is the answer in the local filesystem? \u2192 Read/Grep/Glob
3. Is the answer derivable from context already loaded? \u2192 Just answer
4. Only if 1-3 fail \u2192 Use external tools

Meta-engrams flagged as "[structural transfer \u2014 untested in current domain]" are hypotheses, not rules. Test before applying.`;
}

// src/schemas/meta-engram.ts
import { z as z4 } from "zod";
var StructuralTemplateSchema = z4.object({
  goal_type: z4.string().min(1),
  constraint_type: z4.string().min(1),
  outcome_type: z4.string().min(1),
  template: z4.string().min(1),
  /** Structural frame — declares which relational pattern this meta-engram uses.
   *  Allows flexible analogy beyond rigid [goal]+[constraint]->[outcome]. */
  structure_type: z4.enum([
    "goal-constraint-outcome",
    "feedback-loop",
    "causal-chain",
    "recursive",
    "tradeoff",
    "freeform"
  ]).default("goal-constraint-outcome"),
  /** For patterns that don't fit the standard template fields — the LLM's own structural description */
  freeform_structure: z4.string().optional()
});
var EvidenceEntrySchema = z4.object({
  engram_id: z4.string(),
  domain: z4.string(),
  mapping_rationale: z4.string(),
  alignment_score: z4.number().min(0).max(1)
});
var FalsificationSchema = z4.object({
  expected_conditions: z4.string(),
  expected_exceptions: z4.string(),
  test_prediction: z4.string().optional()
});
var MetaConfidenceSchema = z4.object({
  evidence_count: z4.number().int().min(0),
  domain_count: z4.number().int().min(0),
  structural_depth: z4.number().int().min(1).max(5),
  validation_ratio: z4.number().min(0).max(1).default(0),
  composite: z4.number().min(0).max(1)
});
var DomainCoverageSchema = z4.object({
  validated: z4.array(z4.string()),
  failed: z4.array(z4.string()).default([]),
  predicted: z4.array(z4.string()).default([])
});
var HierarchyPositionSchema = z4.object({
  level: z4.enum(["mop", "top"]),
  parent: z4.string().nullable().default(null),
  children: z4.array(z4.string()).default([])
});
var MetaFieldSchema = z4.object({
  structure: StructuralTemplateSchema,
  evidence: z4.array(EvidenceEntrySchema).min(2),
  domain_coverage: DomainCoverageSchema,
  falsification: FalsificationSchema,
  confidence: MetaConfidenceSchema,
  hierarchy: HierarchyPositionSchema,
  pipeline_version: z4.string(),
  last_validated: z4.string().optional()
});

// src/version-check.ts
var cache = /* @__PURE__ */ new Map();
async function checkForUpdate(packageName, currentVersion, onResult) {
  const cached = cache.get(packageName);
  if (cached) {
    if (onResult) onResult(cached);
    return cached;
  }
  const result = { current: currentVersion, latest: null, updateAvailable: false, checkedAt: null };
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 3e3);
    const res = await fetch(`https://registry.npmjs.org/${packageName}/latest`, {
      signal: controller.signal,
      headers: { Accept: "application/json" }
    });
    clearTimeout(timeout);
    if (!res.ok) {
      cache.set(packageName, result);
      return result;
    }
    const data = await res.json();
    if (!data.version) {
      cache.set(packageName, result);
      return result;
    }
    result.latest = data.version;
    result.updateAvailable = isNewer(data.version, currentVersion);
    result.checkedAt = Date.now();
  } catch {
  }
  cache.set(packageName, result);
  if (onResult) onResult(result);
  return result;
}
function getCachedUpdateCheck(packageName) {
  return cache.get(packageName) ?? null;
}
function clearVersionCache() {
  cache.clear();
}
function isNewer(a, b) {
  const pa = a.split(".").map(Number);
  const pb = b.split(".").map(Number);
  for (let i = 0; i < 3; i++) {
    if ((pa[i] ?? 0) > (pb[i] ?? 0)) return true;
    if ((pa[i] ?? 0) < (pb[i] ?? 0)) return false;
  }
  return false;
}

// src/index.ts
var INGEST_PATTERNS = [
  { re: /(?:we decided|the decision is|agreed to)\s+(.+?)\.?$/gim, type: "architectural" },
  { re: /(?:always|never|must|should)\s+(.+?)\.?$/gim, type: "behavioral" },
  { re: /(?:the convention is|the rule is|the pattern is)\s+(.+?)\.?$/gim, type: "procedural" },
  { re: /(?:use|prefer)\s+(\w+)\s+(?:for|over|instead of)\s+(.+?)\.?$/gim, type: "behavioral" },
  { re: /(?:important|note|remember):\s*(.+?)\.?$/gim, type: "behavioral" }
];
var Plur = class {
  paths;
  config;
  indexedStorage = null;
  _engramCache = /* @__PURE__ */ new Map();
  constructor(options) {
    this.paths = detectPlurStorage(options?.path);
    this.config = loadConfig(this.paths.config);
    if (this.config.index) {
      this.indexedStorage = new IndexedStorage(this.paths.engrams, this.paths.db, this.config.stores);
    }
  }
  /**
   * Load engrams from primary store + all configured stores, with mtime-based caching.
   * Store engram IDs get namespaced: ENG-2026-0401-001 → ENG-DF-2026-0401-001.
   * Primary engrams are returned unchanged.
   */
  _loadAllEngrams() {
    const primary = this._loadCached(this.paths.engrams);
    const stores = this.config.stores ?? [];
    if (stores.length === 0) return primary;
    const all = [...primary];
    for (const store of stores) {
      const storeEngrams = this._loadCached(store.path);
      const prefix = storePrefix(store.scope);
      for (const e of storeEngrams) {
        if (e.scope !== "global" && e.scope !== store.scope && !e.scope.startsWith(store.scope)) {
          logger.debug(`Skipping engram ${e.id} from store ${store.scope}: scope mismatch (${e.scope})`);
          continue;
        }
        const cloned = { ...e };
        if (cloned.scope === "global") {
          cloned.scope = store.scope;
        }
        const originalId = cloned.id;
        cloned.id = cloned.id.replace(/^(ENG|ABS|META)-/, `$1-${prefix}-`);
        cloned._originalId = originalId;
        cloned._storeScope = store.scope;
        all.push(cloned);
      }
    }
    return all;
  }
  /** Load engrams from a path with mtime-based caching */
  _loadCached(path2) {
    let mtime = 0;
    try {
      mtime = fs3.statSync(path2).mtimeMs;
    } catch {
      return [];
    }
    const cached = this._engramCache.get(path2);
    if (cached && cached.mtime === mtime) return cached.engrams;
    const engrams = loadEngrams(path2);
    this._engramCache.set(path2, { mtime, engrams });
    return engrams;
  }
  /** Find which store owns an engram by ID. For namespaced IDs, strips prefix to find in store. */
  _findEngramStore(id) {
    const primaryEngrams = this._loadCached(this.paths.engrams);
    if (primaryEngrams.find((e) => e.id === id)) {
      return { path: this.paths.engrams, readonly: false, originalId: id };
    }
    const stores = this.config.stores ?? [];
    for (const store of stores) {
      const prefix = storePrefix(store.scope);
      const nsPattern = new RegExp(`^(ENG|ABS|META)-${prefix}-`);
      if (nsPattern.test(id)) {
        const originalId = id.replace(nsPattern, "$1-");
        const storeEngrams = this._loadCached(store.path);
        if (storeEngrams.find((e) => e.id === originalId)) {
          return { path: store.path, readonly: store.readonly ?? false, originalId };
        }
      }
    }
    return null;
  }
  /** Create engram, detect conflicts, save. Returns the created engram. */
  learn(statement, context) {
    if (!this.config.allow_secrets) {
      const secrets = detectSecrets(statement);
      if (secrets.length > 0) {
        throw new Error(`Secret detected in statement: ${secrets[0].pattern}. Use config.allow_secrets to override.`);
      }
    }
    return withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const allEngrams = this._loadAllEngrams();
      const id = generateEngramId(allEngrams);
      const scope = context?.scope ?? "global";
      const now = (/* @__PURE__ */ new Date()).toISOString();
      const conflictingEngrams = detectConflicts({ statement, scope }, allEngrams);
      const conflictIds = conflictingEngrams.map((e) => e.id);
      const engram = {
        id,
        version: 2,
        status: "active",
        consolidated: false,
        type: context?.type ?? "behavioral",
        scope,
        visibility: context?.visibility ?? "private",
        statement,
        rationale: context?.rationale,
        source: context?.source,
        domain: context?.domain,
        activation: {
          retrieval_strength: 0.7,
          storage_strength: 1,
          frequency: 0,
          last_accessed: now.slice(0, 10)
        },
        feedback_signals: { positive: 0, negative: 0, neutral: 0 },
        knowledge_anchors: (context?.knowledge_anchors ?? []).map((a) => ({
          path: a.path,
          relevance: a.relevance ?? "supporting",
          snippet: a.snippet
        })),
        associations: [],
        derivation_count: 1,
        tags: context?.tags ?? [],
        pack: null,
        abstract: context?.abstract ?? null,
        derived_from: context?.derived_from ?? null,
        dual_coding: context?.dual_coding,
        polarity: null,
        relations: conflictIds.length > 0 ? {
          broader: [],
          narrower: [],
          related: [],
          conflicts: conflictIds
        } : void 0
      };
      engrams.push(engram);
      saveEngrams(this.paths.engrams, engrams);
      this._syncIndex();
      return engram;
    });
  }
  /**
   * Search engrams, filter by scope/domain/strength, reactivate accessed.
   * Supports two modes:
   *   - 'fast' (default): BM25 keyword search, instant, no API calls
   *   - 'agentic': LLM-assisted semantic search, higher accuracy, requires llm function
   */
  /** Search engrams using fast BM25 keyword matching. Sync, no API calls. */
  recall(query, options) {
    const filtered = this._filterEngrams(options);
    const limit = options?.limit ?? 20;
    const results = searchEngrams(filtered, query, limit);
    this._reactivateResults(results);
    return results;
  }
  /** Search engrams using LLM-assisted semantic filtering. Async, requires llm function. */
  async recallAsync(query, options) {
    const filtered = this._filterEngrams(options);
    const limit = options?.limit ?? 20;
    const results = await agenticSearch(filtered, query, limit, options.llm);
    this._reactivateResults(results);
    return results;
  }
  /** Search engrams using local embeddings (transformers.js). Async, no API calls. */
  async recallSemantic(query, options) {
    const filtered = this._filterEngrams(options);
    const limit = options?.limit ?? 20;
    const results = await embeddingSearch(filtered, query, limit, this.paths.root);
    this._reactivateResults(results);
    return results;
  }
  /** Hybrid search: BM25 + embeddings merged via Reciprocal Rank Fusion. Async, no API calls. */
  async recallHybrid(query, options) {
    const filtered = this._filterEngrams(options);
    const limit = options?.limit ?? 20;
    const results = await hybridSearch(filtered, query, limit, this.paths.root);
    this._reactivateResults(results);
    return results;
  }
  /** Expanded search: LLM query expansion + hybrid search + RRF merge. Opt-in, requires LLM function. */
  async recallExpanded(query, options) {
    const filtered = this._filterEngrams(options);
    const limit = options?.limit ?? 20;
    const results = await expandedSearch(filtered, query, limit, options.llm, this.paths.root);
    this._reactivateResults(results);
    return results;
  }
  /** Get a single engram by ID, regardless of status. Searches primary + all stores. */
  getById(id) {
    const engrams = this._loadAllEngrams();
    return engrams.find((e) => e.id === id) ?? null;
  }
  /** List all active engrams, optionally filtered by scope/domain. No search — returns all matches. */
  list(options) {
    return this._filterEngrams(options);
  }
  /** Filter engrams by scope/domain/strength (shared by both modes) */
  _filterEngrams(options) {
    let engrams;
    if (this.indexedStorage) {
      engrams = this.indexedStorage.loadFiltered({
        status: "active",
        scope: options?.scope,
        domain: options?.domain
      });
    } else {
      engrams = this._loadAllEngrams();
      engrams = engrams.filter((e) => e.status === "active");
      if (options?.domain) {
        engrams = engrams.filter((e) => e.domain?.startsWith(options.domain));
      }
      if (options?.scope) {
        const scope = options.scope;
        engrams = engrams.filter(
          (e) => e.scope === "global" || e.scope === scope || e.scope.startsWith(scope)
        );
      }
    }
    const today = (/* @__PURE__ */ new Date()).toISOString().slice(0, 10);
    engrams = engrams.filter((e) => {
      if (e.temporal?.valid_until && e.temporal.valid_until < today) return false;
      if (e.temporal?.valid_from && e.temporal.valid_from > today) return false;
      return true;
    });
    if (options?.min_strength !== void 0) {
      engrams = engrams.filter((e) => e.activation.retrieval_strength >= options.min_strength);
    }
    return engrams;
  }
  /** Reactivate accessed engrams and update co-access associations */
  _reactivateResults(results) {
    if (results.length === 0) return;
    const isStoreEngram = (e) => e._originalId || /^(ENG|ABS|META)-[A-Z]{3}-/.test(e.id);
    const primaryResults = results.filter((e) => !isStoreEngram(e));
    if (primaryResults.length === 0) return;
    withLock(this.paths.engrams, () => {
      const allEngrams = loadEngrams(this.paths.engrams);
      const resultIds = new Set(primaryResults.map((e) => e.id));
      const today = (/* @__PURE__ */ new Date()).toISOString().slice(0, 10);
      let modified = false;
      for (const e of allEngrams) {
        if (resultIds.has(e.id)) {
          e.activation.retrieval_strength = reactivate(e.activation.retrieval_strength);
          e.activation.last_accessed = today;
          e.activation.frequency += 1;
          modified = true;
        }
      }
      if (results.length >= 2 && this.config.injection?.co_access !== false) {
        const topHalf = results.slice(0, Math.max(2, Math.ceil(results.length / 2)));
        const topIds = topHalf.map((e) => e.id);
        for (const sourceId of topIds) {
          const source = allEngrams.find((e) => e.id === sourceId);
          if (!source) continue;
          for (const targetId of topIds) {
            if (targetId === sourceId) continue;
            const existing = source.associations.find(
              (a) => a.type === "co_accessed" && a.target === targetId
            );
            if (existing) {
              existing.strength = Math.min(0.95, existing.strength + 0.05);
              existing.updated_at = today;
              modified = true;
            } else {
              const coAccessCount = source.associations.filter((a) => a.type === "co_accessed").length;
              if (coAccessCount < 5) {
                source.associations.push({
                  target_type: "engram",
                  target: targetId,
                  type: "co_accessed",
                  strength: 0.3,
                  updated_at: today
                });
                modified = true;
              }
            }
          }
        }
      }
      if (modified) {
        saveEngrams(this.paths.engrams, allEngrams);
        this._syncIndex();
      }
    });
  }
  /** Scored injection within token budget (BM25 only). Returns formatted strings. */
  inject(task, options) {
    return this._formatInjection(task, options);
  }
  /** Scored injection with embedding boost when available. Falls back to BM25 if embeddings not installed. */
  async injectHybrid(task, options) {
    let embeddingBoosts;
    try {
      const engrams = this._loadAllEngrams().filter((e) => e.status === "active");
      const results = await embeddingSearch(engrams, task, engrams.length, this.paths.root);
      if (results.length > 0) {
        embeddingBoosts = /* @__PURE__ */ new Map();
        for (let i = 0; i < results.length; i++) {
          embeddingBoosts.set(results[i].id, 1 / (1 + i * 0.1));
        }
      }
    } catch {
    }
    return this._formatInjection(task, options, embeddingBoosts);
  }
  _formatInjection(task, options, embeddingBoosts) {
    const engrams = this._loadAllEngrams();
    const packs = loadAllPacks(this.paths.packs);
    const budget = options?.budget ?? this.config.injection_budget ?? 2e3;
    const result = selectAndSpread(
      {
        prompt: task,
        scope: options?.scope,
        maxTokens: budget
      },
      engrams,
      packs,
      {
        spread_cap: this.config.injection?.spread_cap,
        spread_budget: this.config.injection?.spread_budget
      },
      embeddingBoosts
    );
    const formatEngrams = (wires) => {
      if (wires.length === 0) return "";
      return wires.map((e) => `[${e.id}] ${e.statement}`).join("\n");
    };
    const directivesStr = formatEngrams(result.directives);
    const constraintsStr = formatEngrams(result.constraints);
    const considerStr = formatEngrams(result.consider);
    const count = result.directives.length + result.constraints.length + result.consider.length;
    const tokensUsed = result.tokens_used.directives + result.tokens_used.consider;
    const injected_ids = [
      ...result.directives.map((e) => e.id),
      ...result.constraints.map((e) => e.id),
      ...result.consider.map((e) => e.id)
    ];
    return {
      directives: directivesStr,
      constraints: constraintsStr,
      consider: considerStr,
      count,
      tokens_used: tokensUsed,
      injected_ids
    };
  }
  /** Update feedback_signals and adjust retrieval_strength. Searches primary, stores, then packs. */
  feedback(id, signal) {
    const found = withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const engram = engrams.find((e) => e.id === id);
      if (!engram) return false;
      if (!engram.feedback_signals) {
        engram.feedback_signals = { positive: 0, negative: 0, neutral: 0 };
      }
      engram.feedback_signals[signal] += 1;
      if (signal === "positive") {
        engram.activation.retrieval_strength = Math.min(1, engram.activation.retrieval_strength + 0.05);
      } else if (signal === "negative") {
        engram.activation.retrieval_strength = Math.max(0, engram.activation.retrieval_strength - 0.1);
      }
      saveEngrams(this.paths.engrams, engrams);
      this._syncIndex();
      return true;
    });
    if (found) return;
    const storeInfo = this._findEngramStore(id);
    if (storeInfo && storeInfo.path !== this.paths.engrams) {
      if (storeInfo.readonly) {
        throw new Error("Engram is in a readonly store");
      }
      const storeEngrams = loadEngrams(storeInfo.path);
      const engram = storeEngrams.find((e) => e.id === storeInfo.originalId);
      if (engram) {
        if (!engram.feedback_signals) {
          engram.feedback_signals = { positive: 0, negative: 0, neutral: 0 };
        }
        engram.feedback_signals[signal] += 1;
        if (signal === "positive") {
          engram.activation.retrieval_strength = Math.min(1, engram.activation.retrieval_strength + 0.05);
        } else if (signal === "negative") {
          engram.activation.retrieval_strength = Math.max(0, engram.activation.retrieval_strength - 0.1);
        }
        saveEngrams(storeInfo.path, storeEngrams);
        this._engramCache.delete(storeInfo.path);
        this._syncIndex();
        return;
      }
    }
    this._feedbackPack(id, signal);
  }
  /** Save extracted meta-engrams to the engram store. Skips IDs that already exist. */
  saveMetaEngrams(metas) {
    return withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const existingIds = new Set(engrams.map((e) => e.id));
      let saved = 0;
      let skipped = 0;
      for (const meta of metas) {
        if (existingIds.has(meta.id)) {
          skipped++;
        } else {
          engrams.push(meta);
          saved++;
        }
      }
      if (saved > 0) {
        saveEngrams(this.paths.engrams, engrams);
        this._syncIndex();
      }
      return { saved, skipped };
    });
  }
  /** Update an existing engram in the store by ID. Returns true if found and updated. */
  updateEngram(updated) {
    return withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const idx = engrams.findIndex((e) => e.id === updated.id);
      if (idx === -1) return false;
      engrams[idx] = updated;
      saveEngrams(this.paths.engrams, engrams);
      this._syncIndex();
      return true;
    });
  }
  /** Set engram status to 'retired'. Supports primary and store engrams. */
  forget(id, reason) {
    const foundInPrimary = withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const engram = engrams.find((e) => e.id === id);
      if (!engram) return false;
      engram.status = "retired";
      if (reason && !engram.rationale) {
        engram.rationale = `Retired: ${reason}`;
      }
      saveEngrams(this.paths.engrams, engrams);
      this._syncIndex();
      return true;
    });
    if (foundInPrimary) return;
    const storeInfo = this._findEngramStore(id);
    if (storeInfo && storeInfo.path !== this.paths.engrams) {
      if (storeInfo.readonly) {
        throw new Error("Cannot retire engram from readonly store");
      }
      const storeEngrams = loadEngrams(storeInfo.path);
      const engram = storeEngrams.find((e) => e.id === storeInfo.originalId);
      if (engram) {
        engram.status = "retired";
        if (reason && !engram.rationale) {
          engram.rationale = `Retired: ${reason}`;
        }
        saveEngrams(storeInfo.path, storeEngrams);
        this._engramCache.delete(storeInfo.path);
        this._syncIndex();
        return;
      }
    }
    throw new Error(`Engram not found: ${id}`);
  }
  /** Remove retired engrams from storage. Returns count of removed and remaining. */
  compact() {
    return withLock(this.paths.engrams, () => {
      const engrams = loadEngrams(this.paths.engrams);
      const active = engrams.filter((e) => e.status !== "retired");
      const removed = engrams.length - active.length;
      if (removed > 0) {
        saveEngrams(this.paths.engrams, active);
        this._syncIndex();
      }
      return { removed, remaining: active.length };
    });
  }
  /** Rebuild SQLite index from YAML source of truth. Only works when index: true. */
  reindex() {
    if (!this.indexedStorage) {
      this.indexedStorage = new IndexedStorage(this.paths.engrams, this.paths.db, this.config.stores);
    }
    this.indexedStorage.reindex();
  }
  /** Sync SQLite index after YAML write (no-op if index disabled) */
  _syncIndex() {
    if (this.indexedStorage) {
      this.indexedStorage.syncFromYaml();
    }
  }
  /** Search packs for an engram by ID and apply feedback, writing back to the pack's engrams.yaml. */
  _feedbackPack(id, signal) {
    if (!fs3.existsSync(this.paths.packs)) throw new Error(`Engram not found: ${id}`);
    for (const entry of fs3.readdirSync(this.paths.packs)) {
      const packDir = `${this.paths.packs}/${entry}`;
      if (!fs3.statSync(packDir).isDirectory()) continue;
      const engramsPath = `${packDir}/engrams.yaml`;
      if (!fs3.existsSync(engramsPath)) continue;
      const engrams = loadEngrams(engramsPath);
      const engram = engrams.find((e) => e.id === id);
      if (!engram) continue;
      if (!engram.feedback_signals) {
        engram.feedback_signals = { positive: 0, negative: 0, neutral: 0 };
      }
      engram.feedback_signals[signal] += 1;
      if (signal === "positive") {
        engram.activation.retrieval_strength = Math.min(1, engram.activation.retrieval_strength + 0.05);
      } else if (signal === "negative") {
        engram.activation.retrieval_strength = Math.max(0, engram.activation.retrieval_strength - 0.1);
      }
      saveEngrams(engramsPath, engrams);
      return;
    }
    throw new Error(`Engram not found: ${id}`);
  }
  /** Capture an episodic memory. */
  capture(summary, context) {
    return captureEpisode(this.paths.episodes, summary, context);
  }
  /** Query the episode timeline. */
  timeline(query) {
    return queryTimeline(this.paths.episodes, query);
  }
  /** Rule-based extraction of engram candidates from content. */
  ingest(content, options) {
    const candidates = [];
    const seen = /* @__PURE__ */ new Set();
    for (const { re, type } of INGEST_PATTERNS) {
      re.lastIndex = 0;
      let match;
      while ((match = re.exec(content)) !== null) {
        const captured = match.slice(1).filter(Boolean).join(" ").trim();
        if (!captured || captured.length < 5) continue;
        if (seen.has(captured.toLowerCase())) continue;
        if (!this.config.allow_secrets && detectSecrets(captured).length > 0) continue;
        seen.add(captured.toLowerCase());
        candidates.push({
          statement: captured,
          type,
          source: options?.source
        });
      }
    }
    if (!options?.extract_only && candidates.length > 0) {
      for (const candidate of candidates) {
        this.learn(candidate.statement, {
          type: candidate.type,
          scope: options?.scope ?? "global",
          domain: options?.domain,
          source: candidate.source
        });
      }
    }
    return candidates;
  }
  /** Install a pack from a source path. Detects conflicts with existing engrams. */
  installPack(source) {
    const existing = this._loadAllEngrams();
    return installPack(this.paths.packs, source, existing);
  }
  /** Uninstall a pack by name. */
  uninstallPack(name) {
    return uninstallPack(this.paths.packs, name);
  }
  /** Export engrams as a shareable pack with privacy scanning and integrity hash. */
  exportPack(engrams, outputDir, manifest) {
    return exportPack(engrams, outputDir, manifest);
  }
  /** List all installed packs (with integrity hashes). */
  listPacks() {
    return listPacks(this.paths.packs);
  }
  /** Sync engrams to git. Initializes repo on first call, commits + push/pull on subsequent calls. */
  sync(remote) {
    return sync(this.paths.root, remote);
  }
  /** Get git sync status without making changes. */
  syncStatus() {
    return getSyncStatus(this.paths.root);
  }
  /** Return system health info. */
  status() {
    const engrams = this._loadAllEngrams();
    const episodes = queryTimeline(this.paths.episodes);
    const packs = listPacks(this.paths.packs);
    return {
      engram_count: engrams.filter((e) => e.status !== "retired").length,
      episode_count: episodes.length,
      pack_count: packs.length,
      storage_root: this.paths.root,
      config: this.config
    };
  }
  /** Register an additional engram store. */
  addStore(storePath, scope, options) {
    const config = loadConfig(this.paths.config);
    const existing = config.stores?.find((s) => s.path === storePath);
    if (existing) return;
    const stores = [...config.stores ?? [], {
      path: storePath,
      scope,
      shared: options?.shared ?? false,
      readonly: options?.readonly ?? false
    }];
    let configData = {};
    try {
      const raw = fs3.readFileSync(this.paths.config, "utf8");
      if (raw) configData = yaml5.load(raw) ?? {};
    } catch {
    }
    configData.stores = stores;
    fs3.writeFileSync(this.paths.config, yaml5.dump(configData, { lineWidth: 120, noRefs: true }));
    this.config = loadConfig(this.paths.config);
  }
  /** List all configured stores. */
  listStores() {
    const stores = this.config.stores ?? [];
    const primary = {
      path: this.paths.engrams,
      scope: "global",
      shared: false,
      readonly: false,
      engram_count: this._loadCached(this.paths.engrams).filter((e) => e.status !== "retired").length
    };
    const additional = stores.map((s) => {
      let count = 0;
      try {
        count = this._loadCached(s.path).filter((e) => e.status !== "retired").length;
      } catch {
      }
      return { ...s, engram_count: count };
    });
    return [primary, ...additional];
  }
};
export {
  DomainCoverageSchema,
  EvidenceEntrySchema,
  FalsificationSchema,
  HierarchyPositionSchema,
  IndexedStorage,
  MetaConfidenceSchema,
  MetaFieldSchema,
  PLATITUDE_PATTERNS,
  Plur,
  SessionBreadcrumbs,
  StructuralTemplateSchema,
  alignCluster,
  analyzeStructure,
  checkForUpdate,
  classifyPolarity,
  clearVersionCache,
  clusterByStructure,
  computeConfidence,
  computeMetaConfidence,
  confidenceBand,
  detectPlurStorage,
  detectSecrets,
  engramSearchText,
  extractMetaEngrams,
  formulateMetaEngram,
  generateGuardrails,
  getCachedUpdateCheck,
  isPlatitude,
  organizeHierarchy,
  tokenSimilarity,
  validateMetaEngram
};
